<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Test a simpler version of veb-boost performance on ukbiobank data-2</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/journal.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">causal-TWAS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/causal-TWAS">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Test a simpler version of veb-boost performance on ukbiobank data-2</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-05-28
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 5 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 2
</p>
<p>
<strong>Knit directory:</strong> <code>causal-TWAS/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191103code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191103)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191103code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191103)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongabsolute"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>File paths:</strong> absolute </a>
</p>
</div>
<div id="strongFilepathsstrongabsolute" class="panel-collapse collapse">
<div class="panel-body">
<p>
Using absolute paths to the files within your workflowr project makes it difficult for you and others to run your code on a different machine. Change the absolute path(s) below to the suggested relative path(s) to make your code more reproducible.
</p>
<table class="table table-condensed table-hover">
<thead>
<tr>
<th style="text-align:left;">
absolute
</th>
<th style="text-align:left;">
relative
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
~/causalTWAS/causal-TWAS/code/fit_mr.ash.R
</td>
<td style="text-align:left;">
code/fit_mr.ash.R
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabcausalTWAStreeb38488a0a6ba62601c65f9ba674c0858e63fed16targetblankb38488aa"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/stephenslab/causal-TWAS/tree/b38488a0a6ba62601c65f9ba674c0858e63fed16" target="_blank">b38488a</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabcausalTWAStreeb38488a0a6ba62601c65f9ba674c0858e63fed16targetblankb38488aa" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .ipynb_checkpoints/
    Ignored:    code/.ipynb_checkpoints/
    Ignored:    data/

Untracked files:
    Untracked:  analysis/figure/
    Untracked:  analysis/simulation-simpleveb-boost2-ukbchr22.Rmd
    Untracked:  analysis/simulation_simpleveb-boost2.Rmd
    Untracked:  analysis/simulation_susie.Rmd
    Untracked:  code/run_test_mr.ash2s.R
    Untracked:  code/run_test_susie.R

Unstaged changes:
    Modified:   analysis/index.Rmd
    Modified:   analysis/simulation_simpleveb-boost.Rmd
    Modified:   code/mr.ash2.R
    Modified:   code/run_test_mr.ash2.R
    Modified:   code/workflow-ashtest.ipynb

Staged changes:
    Renamed1:   code/simple_vebboost.R
    Renamed2:   code/mr.ash2.R
    Renamed1:   code/run_test_veb-boost.R
    Renamed2:   code/run_test_mr.ash2.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<pre class="r"><code>library(mr.ash.alpha)
source(&quot;~/causalTWAS/causal-TWAS/code/fit_mr.ash.R&quot;)

summary_mr.ash &lt;- function(fit){
  cat(&quot;pi1 = &quot;, 1-fit$pi[[1]], &quot;\n&quot;)
  pve &lt;- get_pve(fit)
  cat(&quot;pve = &quot;, pve, &quot;\n&quot;)
}

plot_beta &lt;- function(beta,beta.pm, ...){
  plot( beta, pch=19, col =&quot;darkgreen&quot;, ...)
  points(beta.pm, pch =19, col = &quot;red&quot;)
  legend(&quot;topright&quot;, legend=c(&quot;true beta&quot;, &quot;posterior mean&quot;),
       col=c(&quot;darkgreen&quot;, &quot;red&quot;), pch=19)
}

plot_pip &lt;- function(indi, pip, pipcut = 0.1) {
  x &lt;- barplot(indi, main = &quot;PIP&quot;)
  points(x, pip, pch=19, col=&quot;red&quot;)
  pred &lt;- which(pip &gt; pipcut)
  real &lt;- which(indi==1)
  tp &lt;- length(intersect(pred, real))/length(pred)
  fp &lt;- 1 - tp
  fn &lt;- length(setdiff(real, pred))/length(real)
  cat(&quot;false positive rate: &quot;, fp, &quot;\n&quot;)
  cat(&quot;false negative rate: &quot;, fn, &quot;\n&quot;)
}

summary_mr.ash2 &lt;- function(g.fit, s.fit, phenores, pipcut = 0.1){
  e.b &lt;- rep(0, length(g.fit$beta))
  e.b[phenores$param$idx.cgene] &lt;- phenores$param$e.beta
  cat(&quot;gene expression effect: \n&quot;)
  summary_mr.ash(g.fit)
  plot_beta(e.b, g.fit$beta)
  indi &lt;- rep(0, length(g.fit$beta))
  indi[phenores$param$idx.cgene] &lt;- 1
  pip &lt;- get_pip(g.fit)
  plot_pip(indi, pip, pipcut = pipcut)
  
  s.b &lt;- rep(0, length(s.fit$beta))
  s.b[phenores$param$idx.cSNP] &lt;- phenores$param$s.theta
  cat(&quot;snp effect: \n&quot;)
  summary_mr.ash(s.fit)
  plot_beta(s.b, s.fit$beta)
}</code></pre>
<pre class="r"><code>simdatadir &lt;- &quot;~/causalTWAS/simulations/simulation_ashtest_20200503/&quot;
outputdir &lt;- &quot;~/causalTWAS/simulations/simulation_ashtest_20200527/&quot;</code></pre>
<div id="simulation-1" class="section level2">
<h2>Simulation 1</h2>
<p>True parameters:</p>
<pre class="r"><code>load(paste0(simdatadir, &quot;mr.ash2_20200503-1-pheno.Rd&quot;))
readLines(paste0(simdatadir, &quot;param-20200503-1.R&quot;))</code></pre>
<pre><code>[1] &quot;pve.expr &lt;- 0.01&quot; &quot;pve.snp &lt;- 0.05&quot;  &quot;pi_beta &lt;-  0.1&quot; 
[4] &quot;pi_theta &lt;- 1e-3&quot; &quot;tau &lt;- 1&quot;        </code></pre>
<p>MR.ASH2s results (start from gene):</p>
<pre class="r"><code>load(paste0(outputdir, &quot;20200527-1-mr.ash2s.expr-res.Rd&quot;))
g.fit &lt;- mr.ash2s.fit$fit1
s.fit &lt;-  mr.ash2s.fit$fit2
summary_mr.ash2(g.fit, s.fit, phenores)</code></pre>
<pre><code>gene expression effect: 
pi1 =  0.05417903 
pve =  0.1336998 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%201%20gene-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%201%20gene-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>false positive rate:  0.3 
false negative rate:  0.4615385 
snp effect: 
pi1 =  0.0001887495 
pve =  0.2500488 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%201%20gene-3.png" width="672" style="display: block; margin: auto;" /></p>
<p>MR.ASH results (start from SNP):</p>
<pre class="r"><code>load(paste0(outputdir, &quot;20200527-1-mr.ash2s.snp-res.Rd&quot;))
g.fit &lt;- mr.ash2s.fit$fit2
s.fit &lt;-  mr.ash2s.fit$fit1
summary_mr.ash2(g.fit, s.fit, phenores)</code></pre>
<pre><code>gene expression effect: 
pi1 =  0.05417931 
pve =  0.1337004 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%201%20snp-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%201%20snp-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>false positive rate:  0.3 
false negative rate:  0.4615385 
snp effect: 
pi1 =  0.0001887502 
pve =  0.2500495 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%201%20snp-3.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="simulation-2" class="section level2">
<h2>Simulation 2</h2>
<p>True parameters:</p>
<pre class="r"><code>load(paste0(simdatadir, &quot;mr.ash2_20200503-2-pheno.Rd&quot;))
readLines(paste0(simdatadir, &quot;param-20200503-2.R&quot;))</code></pre>
<pre><code>[1] &quot;pve.expr &lt;- 0.005&quot; &quot;pve.snp &lt;- 0.01&quot;   &quot;pi_beta &lt;-  0.1&quot;  
[4] &quot;pi_theta &lt;- 1e-3&quot;  &quot;tau &lt;- 1&quot;         </code></pre>
<p>MR.ASH results (start from gene):</p>
<pre class="r"><code>load(paste0(outputdir, &quot;20200527-2-mr.ash2s.expr-res.Rd&quot;))
g.fit &lt;- mr.ash2s.fit$fit1
s.fit &lt;-  mr.ash2s.fit$fit2
summary_mr.ash2(g.fit, s.fit, phenores)</code></pre>
<pre><code>gene expression effect: 
pi1 =  1.973726e-10 
pve =  7.959401e-08 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%202%20gene-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%202%20gene-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>false positive rate:  NaN 
false negative rate:  1 
snp effect: 
pi1 =  1.115821e-07 
pve =  0.0002921814 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%202%20gene-3.png" width="672" style="display: block; margin: auto;" /></p>
<p>MR.ASH results (start from SNP):</p>
<pre class="r"><code>load(paste0(outputdir, &quot;20200527-2-mr.ash2s.snp-res.Rd&quot;))
g.fit &lt;- mr.ash2s.fit$fit2
s.fit &lt;-  mr.ash2s.fit$fit1
summary_mr.ash2(g.fit, s.fit, phenores)</code></pre>
<pre><code>gene expression effect: 
pi1 =  1.966005e-10 
pve =  7.958468e-08 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%202%20snp-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%202%20snp-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>false positive rate:  NaN 
false negative rate:  1 
snp effect: 
pi1 =  1.196675e-07 
pve =  0.0003063728 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%202%20snp-3.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="simulation-3" class="section level2">
<h2>Simulation 3</h2>
<p>True parameters:</p>
<pre class="r"><code>load(paste0(simdatadir, &quot;mr.ash2_20200503-3-pheno.Rd&quot;))
readLines(paste0(simdatadir, &quot;param-20200503-3.R&quot;))</code></pre>
<pre><code>[1] &quot;pve.expr &lt;- 0.01&quot; &quot;pve.snp &lt;- 0.05&quot;  &quot;pi_beta &lt;-  0.05&quot;
[4] &quot;pi_theta &lt;- 1e-4&quot; &quot;tau &lt;- 1&quot;        </code></pre>
<p>MR.ASH results (start from gene):</p>
<pre class="r"><code>load(paste0(outputdir, &quot;20200527-3-mr.ash2s.expr-res.Rd&quot;))
g.fit &lt;- mr.ash2s.fit$fit1
s.fit &lt;-  mr.ash2s.fit$fit2
summary_mr.ash2(g.fit, s.fit, phenores)</code></pre>
<pre><code>gene expression effect: 
pi1 =  0.02753678 
pve =  0.07367684 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%203%20gene-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%203%20gene-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>false positive rate:  0.3333333 
false negative rate:  0.7142857 
snp effect: 
pi1 =  0.0001530549 
pve =  0.2156009 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%203%20gene-3.png" width="672" style="display: block; margin: auto;" /></p>
<p>MR.ASH results (start from SNP):</p>
<pre class="r"><code>load(paste0(outputdir, &quot;20200527-3-mr.ash2s.snp-res.Rd&quot;))
g.fit &lt;- mr.ash2s.fit$fit2
s.fit &lt;-  mr.ash2s.fit$fit1
summary_mr.ash2(g.fit, s.fit, phenores)</code></pre>
<pre><code>gene expression effect: 
pi1 =  0.009105761 
pve =  0.02562701 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%203%20snp-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%203%20snp-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>false positive rate:  0 
false negative rate:  0.8571429 
snp effect: 
pi1 =  0.0001684206 
pve =  0.2321756 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%203%20snp-3.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="simulation-4" class="section level2">
<h2>Simulation 4</h2>
<p>True parameters:</p>
<pre class="r"><code>load(paste0(simdatadir, &quot;mr.ash2_20200503-4-pheno.Rd&quot;))
readLines(paste0(simdatadir, &quot;param-20200503-4.R&quot;))</code></pre>
<pre><code>[1] &quot;pve.expr &lt;- 0.005&quot; &quot;pve.snp &lt;- 0.01&quot;   &quot;pi_beta &lt;-  0.05&quot; 
[4] &quot;pi_theta &lt;- 1e-4&quot;  &quot;tau &lt;- 1&quot;         </code></pre>
<p>MR.ASH results (start from gene):</p>
<pre class="r"><code>load(paste0(outputdir, &quot;20200527-4-mr.ash2s.expr-res.Rd&quot;))
g.fit &lt;- mr.ash2s.fit$fit1
s.fit &lt;-  mr.ash2s.fit$fit2
summary_mr.ash2(g.fit, s.fit, phenores)</code></pre>
<pre><code>gene expression effect: 
pi1 =  0.03483871 
pve =  0.09146919 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%204%20gene-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%204%20gene-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>false positive rate:  0.5 
false negative rate:  0.7142857 
snp effect: 
pi1 =  4.177662e-05 
pve =  0.0699035 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%204%20gene-3.png" width="672" style="display: block; margin: auto;" /></p>
<p>MR.ASH results (start from SNP):</p>
<pre class="r"><code>load(paste0(outputdir, &quot;20200527-4-mr.ash2s.snp-res.Rd&quot;))
g.fit &lt;- mr.ash2s.fit$fit2
s.fit &lt;-  mr.ash2s.fit$fit1
summary_mr.ash2(g.fit, s.fit, phenores)</code></pre>
<pre><code>gene expression effect: 
pi1 =  0.02582045 
pve =  0.06943356 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%204%20snp-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%204%20snp-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre><code>false positive rate:  0.3333333 
false negative rate:  0.7142857 
snp effect: 
pi1 =  5.635395e-05 
pve =  0.09201309 </code></pre>
<p><img src="figure/simulation-simpleveb-boost2-ukbchr22.Rmd/sim%204%20snp-3.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.1 (2018-07-02)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.2.19-el7-x86_64/lib/libopenblas_haswellp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] mr.ash.alpha_0.1-7

loaded via a namespace (and not attached):
 [1] workflowr_1.6.0 Rcpp_1.0.4.6    lattice_0.20-38 digest_0.6.18  
 [5] later_0.7.5     rprojroot_1.3-2 grid_3.5.1      R6_2.3.0       
 [9] backports_1.1.2 git2r_0.26.1    magrittr_1.5    evaluate_0.12  
[13] highr_0.7       stringi_1.3.1   fs_1.3.1        promises_1.0.1 
[17] Matrix_1.2-15   rmarkdown_1.10  tools_3.5.1     stringr_1.4.0  
[21] glue_1.3.0      httpuv_1.4.5    yaml_2.2.0      compiler_3.5.1 
[25] htmltools_0.3.6 knitr_1.20     </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
