<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Simulation-WTCCC-2</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/journal.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">causal-TWAS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/causal-TWAS">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Simulation-WTCCC-2</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-04-04
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>causal-TWAS/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191103code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191103)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191103code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191103)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabcausalTWAStree2bc1493ba6245fd1df6052ac07db3d6f4e8ff3abtargetblank2bc1493a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/stephenslab/causal-TWAS/tree/2bc1493ba6245fd1df6052ac07db3d6f4e8ff3ab" target="_blank">2bc1493</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabcausalTWAStree2bc1493ba6245fd1df6052ac07db3d6f4e8ff3abtargetblank2bc1493a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .ipynb_checkpoints/
    Ignored:    code/.ipynb_checkpoints/
    Ignored:    data/

Unstaged changes:
    Modified:   analysis/simulation-WTCCC.Rmd
    Modified:   analysis/simulation-WTCCC2.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/causal-TWAS/blob/f43fd469d91652cf0fef798d3b43ceed63507282/analysis/simulation-WTCCC2.Rmd" target="_blank">f43fd46</a>
</td>
<td>
simingz
</td>
<td>
2020-03-24
</td>
<td>
fix results display bug
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/causal-TWAS/f43fd469d91652cf0fef798d3b43ceed63507282/docs/simulation-WTCCC2.html" target="_blank">f43fd46</a>
</td>
<td>
simingz
</td>
<td>
2020-03-24
</td>
<td>
fix results display bug
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/causal-TWAS/blob/5a4a5888be36e6455675aa81582d928eb6d4e084/analysis/simulation-WTCCC2.Rmd" target="_blank">5a4a588</a>
</td>
<td>
simingz
</td>
<td>
2020-03-13
</td>
<td>
v1.0 simulate expression using separate data 20191127
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/causal-TWAS/5a4a5888be36e6455675aa81582d928eb6d4e084/docs/simulation-WTCCC2.html" target="_blank">5a4a588</a>
</td>
<td>
simingz
</td>
<td>
2020-03-13
</td>
<td>
v1.0 simulate expression using separate data 20191127
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Use WTCCC data obtained from Peter on CRI.</p>
<p>The difference between this analysis and <a href="simulation-WTCCC.html">previous run</a>: used 500 samples from <code>cad.RData</code> from WTCCC to simulate expression, then train expression model using <code>cv.glmnet</code>. We impute expression in using <code>bd.RData</code> from WTCCC.</p>
<p>Simulation of phenotype: use the same weights as used in simulate expression using the <code>'cad.RData</code> dataset, simulate phenotype according to:</p>
<p><span class="math display">\[Y = G\alpha\gamma + G\theta + \epsilon\]</span>. For <span class="math inline">\(\alpha\)</span>, we used the same as used in simulating expression (Thus, not <span class="math inline">\(\tilde{X}\)</span>). <span class="math inline">\(\gamma \sim N(0,\sigma_\gamma^2), \theta \sim N(0, \sigma_\theta^2), \epsilon \sim N (0,1)\)</span>.</p>
<p>The genotype data and other settings are the same.</p>
<div id="simulate-expression" class="section level2">
<h2>Simulate expression</h2>
<p>We simulate expression based on the following model, using 500 samples from cad in WTCCC:</p>
<p><span class="math display">\[ X = G\alpha + \xi\]</span> For each gene, we sample <span class="math inline">\(L\)</span> eQTLs for this gene, for eQTL <span class="math inline">\(k\)</span>, we have <span class="math display">\[ \alpha_k \sim N(0,\sigma_\alpha^2) \]</span> <span class="math display">\[ \xi \sim N(0,1) \]</span> Then we have the heritability of the gene:</p>
<p><span class="math display">\[\begin{aligned}
h^2_{eQTL} &amp;= \frac{var(G\alpha)}{var(G\alpha)+var(\xi)} \\
&amp;= \frac{\Sigma_kvar(G_k) \alpha_k^2}{\Sigma_kvar(G_k) \alpha_k^2 + var(\xi)}\\
&amp;=  \frac{\Sigma_k\alpha_k^2}{\Sigma_k\alpha_k^2 + var(\xi)}\\
&amp;\approx \frac{\Sigma_k\sigma_\alpha^2}{\Sigma_k\sigma_\alpha^2 + var(\xi)} \\
&amp;=\frac{K\sigma_\alpha^2}{1+K\sigma_\alpha^2}
\end{aligned}\]</span></p>
<p>Here, we use scaled genotype data, so <span class="math inline">\(var(G)=1\)</span>. We also have <span class="math inline">\(\alpha^2_k \approx E(\alpha_k^2)=var(\alpha_k^2)=\sigma_\alpha^2\)</span>. Usually, <span class="math inline">\(K\)</span> ranges from 1 to 5, let <span class="math inline">\(\sigma_\alpha = 0.3\)</span>, then <span class="math inline">\(h^2_{eQTL}\)</span> ranges from 0.08 to 0.31, this is consistent with gene cis-heritability in reality.</p>
</div>
<div id="simulate-quantitative-phenotype" class="section level2">
<h2>Simulate quantitative phenotype</h2>
<p>We simulate phenotype using bd data from WTCCC following <span class="math display">\[Y = G\alpha\gamma + G\theta + \epsilon\]</span>. For <span class="math inline">\(\alpha\)</span>, we used the same as used in simulating expression. <span class="math inline">\(\gamma \sim N(0,\sigma_\gamma^2), \theta \sim N(0, \sigma_\theta^2), \epsilon \sim N (0,1)\)</span>.</p>
<p><span class="math display">\[\begin{align}
PVE_{SNP} = &amp;\frac{var(G\theta)}{var(G\theta) + var(\tilde{X}\gamma) + \sigma_e^2}\\ 
\approx &amp;\frac{M\sigma_\theta^2}{M\sigma_\theta^2 + \Sigma_jvar(\tilde{X_j})\gamma_j^2+ \sigma_e^2}\\ 
\approx &amp;\frac{M\sigma_\theta^2}{M\sigma_\theta^2 + Jvar(\tilde{X})\sigma_\gamma^2+ \sigma_e^2}\\   
\end{align}\]</span></p>
<p><span class="math display">\[
PVE_{expr} \approx \frac{Jvar(\tilde{X})\sigma_\gamma^2}{M\sigma_\theta^2 + Jvar(\tilde{X})\sigma_\gamma^2+ \sigma_e^2}\]</span></p>
<p>Here, <span class="math inline">\(var(\tilde{X})\)</span> is the cis-heratbility of gene expression. <span class="math inline">\(M\)</span> and <span class="math inline">\(J\)</span> are numbers of causal SNP and gene respectively.</p>
<p>Thus in order to get desired <span class="math inline">\(PVE_{SNP}\)</span> and <span class="math inline">\(PVE_{expr}\)</span>, we set <span class="math inline">\(\sigma_\theta^2\)</span> and <span class="math inline">\(\sigma_\gamma^2\)</span> based on the following formula:</p>
<p><span class="math display">\[\sigma_\theta^2 = \frac{PVE_{SNP}}{M(1-PVE_{SNP} - PVE_{expr})}\]</span> <span class="math display">\[\sigma_\gamma^2 = \frac{PVE_{expr}}{Jvar(\tilde{X})(1-PVE_{SNP} - PVE_{expr})}\]</span></p>
<pre class="r"><code>simudir &lt;- &quot;/home/simingz/causalTWAS/simulations/simulation_WTCCC_20191127/&quot;
load(&quot;/home/simingz/causalTWAS/WTCCC/bd.toy.RData&quot;)
Gvar &lt;- mean(apply(X,2,var))
show_res &lt;- function(simutag){
  load(Sys.glob(paste0(simudir,simutag,&quot;*phenotype.Rd&quot;)))
  outdf1 &lt;- data.frame(truth = c(sigma_gamma^2*J.c, sigma_theta^2*M.c/Gvar , 1))
  row.names(outdf1) &lt;- c(&quot;sigma_gamma^2&quot;,&quot;sigma_theta^2&quot;,&quot;sigma_e^2&quot;)
  res &lt;- readLines(paste0(simudir, simutag, &quot;_VC/output/&quot;, simutag, &quot;.log.txt&quot;))
  outdf1$est &lt;- as.numeric(strsplit(res[24], &quot;  &quot;)[[1]][2:4])
  outdf1$est.se &lt;- as.numeric(strsplit(res[25], &quot;  &quot;)[[1]][2:4])
  print(outdf1)
  outdf2 &lt;- data.frame(&quot;est&quot;= c(as.numeric(strsplit(res[20], &quot;  &quot;)[[1]][2:3]), as.numeric(strsplit(res[22], &quot;=&quot;)[[1]][2])))
  row.names(outdf2) &lt;- c(&quot;PVE.expr&quot;,&quot;PVE.snp&quot;,&quot;PVE.total&quot;)
  outdf2$est.se &lt;- c(as.numeric(strsplit(res[21], &quot;  &quot;)[[1]][2:3]), as.numeric(strsplit(res[23], &quot;=&quot;)[[1]][2]))
  print(outdf2)
}</code></pre>
</div>
<div id="simulation-1-m-no.-of-all-snps-j-no.-all-genes" class="section level2">
<h2>Simulation 1: <span class="math inline">\(M\)</span> = No. of all SNPs, <span class="math inline">\(J\)</span> = No. all genes</h2>
<p>We simulate quantitative phenotype under several scenarios. First we make all genes with imputed gene expression as causal genes and all SNPs as causal SNPs. This matches our polygenic version of the model.</p>
<div id="pve_snp-approx-0.1-pve_expr-approx-0.1" class="section level3">
<h3>1.1 <span class="math inline">\(PVE_{SNP} \approx 0.1\)</span>, <span class="math inline">\(PVE_{expr} \approx 0.1\)</span></h3>
<pre class="r"><code>show_res(&quot;S1.1&quot;)</code></pre>
<pre><code>                  truth         est     est.se
sigma_gamma^2 0.5451204 0.000225689 0.00021374
sigma_theta^2 0.4365506 0.872810000 0.23609300
sigma_e^2     1.0000000 1.018880000 0.07640700
                 est     est.se
PVE.expr  0.00412807 0.00390952
PVE.snp   0.21549800 0.05829160
PVE.total 0.21962600 0.05852090</code></pre>
</div>
<div id="pve_snpapprox-0.4-pve_exprapprox-0.1" class="section level3">
<h3>1.2 <span class="math inline">\(PVE_{SNP}\approx 0.4\)</span>, <span class="math inline">\(PVE_{expr}\approx 0.1\)</span></h3>
<pre class="r"><code>show_res(&quot;S1.2&quot;)</code></pre>
<pre><code>                  truth         est      est.se
sigma_gamma^2 0.8721927 0.000321793 0.000292108
sigma_theta^2 2.7939238 2.839750000 0.427679000
sigma_e^2     1.0000000 1.153150000 0.138334000
                 est     est.se
PVE.expr  0.00370128 0.00335983
PVE.snp   0.44090200 0.06640170
PVE.total 0.44460300 0.06662640</code></pre>
</div>
</div>
<div id="simulation-2-m-no.-of-all-snps-j-10-all-genes-j2000" class="section level2">
<h2>Simulation 2: <span class="math inline">\(M\)</span> = No. of all SNPs, <span class="math inline">\(J\)</span> = ~10% all genes (J=2000)</h2>
<div id="pve_snp-approx-0.1-pve_expr-approx-0.1-1" class="section level3">
<h3>2.1 <span class="math inline">\(PVE_{SNP} \approx 0.1\)</span>, <span class="math inline">\(PVE_{expr} \approx 0.1\)</span></h3>
<pre class="r"><code>show_res(&quot;S2.1&quot;)</code></pre>
<pre><code>                  truth         est      est.se
sigma_gamma^2 0.5451204 0.000111696 0.000139882
sigma_theta^2 0.4365506 0.528777000 0.208465000
sigma_e^2     1.0000000 1.060910000 0.067136400
                 est     est.se
PVE.expr  0.00216157 0.00270702
PVE.snp   0.13813100 0.05445660
PVE.total 0.14029200 0.05440400</code></pre>
</div>
<div id="pve_snp-approx-0.4-pve_expr-approx-0.1" class="section level3">
<h3>2.2 <span class="math inline">\(PVE_{SNP} \approx 0.4\)</span>, <span class="math inline">\(PVE_{expr} \approx 0.1\)</span></h3>
<pre class="r"><code>show_res(&quot;S2.2&quot;)</code></pre>
<pre><code>                  truth         est      est.se
sigma_gamma^2 0.8721927 0.000258743 0.000278473
sigma_theta^2 2.7939238 2.144150000 0.372914000
sigma_e^2     1.0000000 1.301840000 0.120221000
                 est     est.se
PVE.expr  0.00309079 0.00332647
PVE.snp   0.34573400 0.06013060
PVE.total 0.34882500 0.06013430</code></pre>
</div>
</div>
<div id="simulation-3-m-1-all-snps-m5000-j-no.-all-genes" class="section level2">
<h2>Simulation 3: <span class="math inline">\(M\)</span> = ~ 1% all SNPs (M=5000), <span class="math inline">\(J\)</span> = No. all genes</h2>
<div id="pve_snp-approx-0.1-pve_expr-approx-0.1-2" class="section level3">
<h3>3.1 <span class="math inline">\(PVE_{SNP} \approx 0.1\)</span>, <span class="math inline">\(PVE_{expr} \approx 0.1\)</span></h3>
<pre class="r"><code>show_res(&quot;S3.1&quot;)</code></pre>
<pre><code>                  truth         est      est.se
sigma_gamma^2 0.5451204 8.28456e-06 7.35267e-05
sigma_theta^2 0.4365506 5.06254e-01 2.20805e-01
sigma_e^2     1.0000000 1.10084e+00 7.12064e-02
                  est     est.se
PVE.expr  0.000156494 0.00138891
PVE.snp   0.129088000 0.05630220
PVE.total 0.129244000 0.05632350</code></pre>
</div>
<div id="pve_snp-approx-0.4-pve_expr-approx-0.1-1" class="section level3">
<h3>3.2 <span class="math inline">\(PVE_{SNP} \approx 0.4\)</span>, <span class="math inline">\(PVE_{expr} \approx 0.1\)</span></h3>
<pre class="r"><code>show_res(&quot;S3.2&quot;)</code></pre>
<pre><code>                  truth         est      est.se
sigma_gamma^2 0.8721927 0.000185945 0.000246806
sigma_theta^2 2.7939238 2.467070000 0.407404000
sigma_e^2     1.0000000 1.227570000 0.131413000
                est     est.se
PVE.expr  0.0021904 0.00290734
PVE.snp   0.3922910 0.06478150
PVE.total 0.3944810 0.06482130</code></pre>
</div>
</div>
<div id="discussion" class="section level2">
<h2>Discussion</h2>
<p>In this analysis, we see the PVE estimation for expression is much under estimated, this is because many of the gene expression training model has low <span class="math inline">\(R^2\)</span>, especially for low heritability ones. For a polygenic model, because we are estimating the total herability explained by expression, the ones with low <span class="math inline">\(R^2\)</span> will affect this estimation. We are moving to model with a sparse prior for expression.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.1 (2018-07-02)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Scientific Linux 7.4 (Nitrogen)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.2.19-el7-x86_64/lib/libopenblas_haswellp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] workflowr_1.6.0 Rcpp_1.0.0      digest_0.6.18   later_0.7.5    
 [5] rprojroot_1.3-2 R6_2.3.0        backports_1.1.2 git2r_0.26.1   
 [9] magrittr_1.5    evaluate_0.12   stringi_1.3.1   fs_1.3.1       
[13] promises_1.0.1  whisker_0.3-2   rmarkdown_1.10  tools_3.5.1    
[17] stringr_1.4.0   glue_1.3.0      httpuv_1.4.5    yaml_2.2.0     
[21] compiler_3.5.1  htmltools_0.3.6 knitr_1.20     </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
