
# Analysis description

```{r param}
n <- 20000 # number of samples
p <- 86000 # number of SNPs
```

The genotype data we used is from UKB biobank, randomly selecting n=`r as.character(n)` samples. We use SNP genotype data from chr 17 to chr 22 combined. These genomic regions represents 12.5% of the genome. SNPs are downsampled to 1/10 (randomly), eQTLs (see below for definition of eQTL) were added back. This ends up with p = `p as.charater(p)` SNPs.

Our analysis consists of the following steps: 

1. Build expression predictors using another expression-genotype dataset. 

The one we used in this analysis is GTEx Adipose tissue v7 dataset. This dataset contains ~ 380 samples. FUSION/TWAS were used train expression model and we used their lasso results. There are around 1700 genes with expression  