
# Analysis description

```{r param}
n <- 20000 # number of samples
p <- 86000 # number of SNPs
J <- 1653 # number of genes
```

The genotype data we used is from UKB biobank, randomly selecting n = `r as.character(n)` samples. We use SNP genotype data from chr 17 to chr 22 combined. These genomic regions represents 12.5% of the genome. SNPs are downsampled to 1/10 (randomly), eQTLs (see below for definition of eQTL) were added back. This ends up with p = `p as.charater(p)` SNPs.

Our analysis consists of the following steps: 

1. Build expression predictors using another expression-genotype dataset. 

The one we used in this analysis is GTEx Adipose tissue v7 dataset. This dataset contains ~ 380 samples. FUSION/TWAS were used to train expression model and we used their lasso results. eQTLs are defined as SNPs with abs(effectize) > 1e-8.  There are `r as.character(J)` genes with expression from chr17 to chr22. 

2. Run susie iteratively for regions.

We impute gene expression for our genotype data. Next, the analysis is done at the “region” level, which is about 500kb bins along the genome. Each bin would contain all the SNPs, as well as all the genes in that bin. The “genes” here mean the imputed expression from genotypes using the expression predictors.
10:18 AM
So the features of SuSiE are: SNPs and imputed genes (not cis-eQTLs of that gene). (edited) 
10:22 AM
Actually I just realize that there may be potential issues here: the cis-eQTLs of the gene may be outside the bins containing the gene. In that case, if eQTL is in LD with a causal SNP in another bin, SuSiE may still find that gene as causal, which is false.
10:22 AM
This may not be common though, if the bin is relatively large.
10:23 AM
Also to give an idea of the size of bin: on average, it contains about 7 genes and 120 SNPs in Siming’s simualation.