---
title: "Visualize ukb whole genome, s40.22"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message= F, warning = F)
```
Provide visualization for analysis [here]

```{r output dir}
simdatadir <- "~/causalTWAS/simulations/simulation_ashtest_20201001/"
outputdir <- "~/causalTWAS/simulations/simulation_susieI_20201001/"

source("analysis/visual_twas_plots.R")
```

```{r functions file}

plot_all <- function(SERf, phenof, chrom, ldres = NULL){
  a <- fread(SERf , header =T)
  load(phenof)

  a <- a[a$chr == chrom,]
  
  if (!is.null(ldres)){
    ld <- read.table(ldres, header =T)
    a <- merge(a, ld, by = "name", all.x = T)}
  
  a <- a[, ifcausal:=as.character(ifcausal)]
  a[a$ifcausal == "1", "ifcausal"] = "Causal"
  a[a$ifcausal == "0", "ifcausal"] = "Non causal"
  
  colnames(a)[colnames(a) == "SERpip"] <-  "susiePIP"
  colnames(a)[colnames(a) == "pos"] <-  "p0"
  
  a[, "PVALUE"] <- -log10(a[, "pvalue"])
  a[, "lbf"] <- log10(exp(a[, "lbf"]))
  
  dt.gene <- a[a$type == "gene", ]
  dt.snp <- a[a$type == "SNP", ]
  
  fig1 <- plot_susie_pip(dt.gene) 
  fig2 <- plot_lbf(dt.gene)
  fig3 <- plot_p(dt.gene) 

  fig4 <- plot_susie_pip(dt.snp)
  fig5 <- plot_lbf(dt.snp) 
  fig6 <- plot_p(dt.snp) %>% layout( plot_bgcolor='#EFF6FC')
  
  fig <- subplot(fig1, fig2, fig3, fig4, fig5, fig6, nrows=6, shareX = T, titleY = T)
  
  fig <- fig %>% layout(autosize = F, width = 600, height = 800, margin = 0.02,
                        xaxis = list(title = "Genomic position"))
  
  fig <- fig %>% toWebGL()
  fig
}
```

## Simulation- high power

First three rows are for genes, last three rows are for SNPs.

```{r sim 2-1-1, out.width = "98%", out.height="95%"}
chrom <- 1
tag <- "20201001-2-1"
pipfile <-  paste0(outputdir, tag, ".config2.1.SERI.txt")
phenof <- paste0(simdatadir, "simu_", tag, "-pheno.Rd")
plot_all(pipfile, phenof, chrom)
```

