---
title: "Test susieI using whole genome, 4k ukb samples"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,  message= F, warning = F)
```
```{r libs}
library(mr.ash.alpha)
library(data.table)
suppressMessages({library(plotly)})
library(tidyr)
library(plyr)
library(stringr)
library(kableExtra)
source("analysis/summarize_twas_plots.R")
```

# Analysis description

```{r parameters}
n <- 40000 # number of samples
p <- 656321 # number of SNPs
J <- 8021 # number of genes
```

The genotype data we used is from UKB biobank, randomly selecting n = `r as.character(n)` samples. We use SNP genotype data from chr 1 to chr 22 combined from UKB. SNPs are downsampled to 1/10 (randomly), eQTLs (see below for definition of eQTL) were added back. This ends up with p = `p as.charater(p)` SNPs.

Our analysis consists of the following steps: 

1. Build expression predictors using another expression-genotype dataset. 

The one we used in this analysis is GTEx Adipose tissue v7 dataset. This dataset contains ~ 380 samples. [FUSION/TWAS](http://gusevlab.org/projects/fusion/) were used to train expression model and we used their lasso results. SNPs included in eQTL anlaysis are restricted to cis-locus 500kb on either side of the gene boundary. eQTLs are defined as SNPs with abs(effectize) > 1e-8 in lasso results.  

2. Impute expression. 

We impute gene expression for our genotype data using expression models obtained from step 1. There are `r as.character(J)` genes with expression model. We imputed expression from genotypes using the expression predictors. 

3. Define and select regions

Next, the analysis is done at the “region” level, which is 500kb bins along the genome. Each bin would contain all the SNPs, as well as all the genes in that bin. We are exploring several ways to select regions that contain true signals, e.g. based on regional sum of mr.ash PIP for genes/SNPs, region smallest TWAS p value for gene/SNPs, or regional bayes factors, etc.

4. Run susie iteratively
We then run susie for each of these regions. So the features of SuSiE are: SNPs and "genes" (not cis-eQTLs of that gene). We use the same prior for all SNPs and another prior for all "genes" when running SUSIE. In some settings, we also run SUSIE with null weight, which is calculated as `1- prior.SNP * n.SNP - prior.gene * n.gene`. We obtain the PIP for SNPs and gene in the region. After we run susie for all regions (one iteration), we take the average of all SNP PIPs as the prior of SNPs for the next iteration and similarly for the prior for genes. 

5. We obtain PIP for genes from the last iteration as results.


```{r dirs}
simdatadir <- "~/causalTWAS/simulations/simulation_ashtest_20201001/"
outputdir <- "~/causalTWAS/simulations/simulation_susieI_20201001/"
susiedir <- "~/causalTWAS/simulations/simulation_susieI_20201001/"
simtag <- '20201001-1-3'
source('~/causalTWAS/causal-TWAS/code/gwas.R')
source('~/causalTWAS/causal-TWAS/code/ld.R')
```

```{r causal}
exprgwasf <- paste0(simdatadir, simtag, ".exprgwas.txt.gz")
load(paste0(simdatadir, "simu_", simtag, "-pheno.Rd"))
caulist <- list()
for (chrom in 1:22) {
  load(paste0("~/causalTWAS/ukbiobank/ukb_chr", chrom ,"_s40000.FBM.Rd"))
  load(paste0(simdatadir, "simu_s40000_GTEXadipose-B", chrom, "-cis-expr.Rd"))
  caulist[[chrom]]<- c(exprres$gnames[phenores$batch[[chrom]]$param$idx.cgene], dat$snp[phenores$batch[[chrom]]$param$idx.cSNP,])
}
cau <- unlist(caulist)
```

# p value distribution

* TWAS p value of genes:

```{r gene twas, fig.height=8, fig.width=8}
chrom <- 22
a <- read.table(exprgwasf, header = T)
a$ifcausal <- ifelse(a$MARKER_ID %in% cau, 1, 0)
ax <- pretty(0:max(-log10(a$PVALUE)), n = 30)

par(mfrow=c(3,1))
h1 <- hist(-log10(a$PVALUE), breaks = 100, xlab = "-log10(p)", main = "P value distribution-all", col = "grey"); grid()
h2 <- hist(-log10(a[a$ifcausal == 1, ]$PVALUE), breaks = h1$breaks, xlab = "-log10(p)", main = "P value distribution-causal", col = "salmon");grid()

plot(a[a$X.CHROM ==chrom, ]$BEGIN, -log10(a[a$X.CHROM ==chrom, ]$PVALUE), col = a[a$X.CHROM ==chrom, ]$ifcausal + 1, xlab = paste0("chr", chrom), ylab = "-log10(pvalue)")
points(a[a$X.CHROM ==chrom & a$ifcausal ==1, ]$BEGIN, -log10(a[a$X.CHROM ==chrom & a$ifcausal ==1, ]$PVALUE), col = "red", pch =19)
grid()
```

* TWAS p value of SNPs:

```{r snp twas, fig.height=8, fig.width=8}
snpgwasf <- paste0(simdatadir, simtag, ".snpgwas.txt.gz")

b <- fread(snpgwasf, header = T)
b$ifcausal <- ifelse(b$MARKER_ID %in% cau, 1, 0)
ax <- pretty(0:max(-log10(b$PVALUE)), n = 30)

par(mfrow=c(3,1))
h1 <- hist(-log10(b$PVALUE), breaks = 100, xlab = "-log10(p)", main = "P value distribution-all", col = "grey"); grid()
h2 <- hist(-log10(b[b$ifcausal == 1, ]$PVALUE), breaks = h1$breaks, xlab = "-log10(p)", main = "P value distribution-causal", col = "salmon");grid()

plot(b[b$X.CHROM ==chrom, ]$BEGIN, -log10(b[b$X.CHROM ==chrom, ]$PVALUE), col = b[b$X.CHROM ==chrom,]$ifcausal + 1, xlab = paste0("chr", chrom), ylab = "-log10(pvalue)")
points(b[b$X.CHROM ==chrom & b$ifcausal ==1, ]$BEGIN, -log10(b[b$X.CHROM ==chrom & b$ifcausal ==1, ]$PVALUE), col = "red", pch =19)
grid()
```

# Exploration of unusual small p values
Take chromosom 22 geno type and perform linear regression for against geno type for causal SNPs on other chromosomes. We see some very significant p values. Take chromosome 2 causal SNPs as an example.


## normal distribution as phenotype
```{r random, fig.height=4, fig.width=8}
chrom <- 22
load(paste0("~/causalTWAS/ukbiobank/ukb_chr", chrom ,"_s40000.FBM.Rd"))
load(paste0(simdatadir, "simu_s40000_GTEXadipose-B", chrom, "-cis-expr.Rd"))
load(paste0(simdatadir, "simu_", simtag, "-pheno.Rd"))

N<- nrow(phenores$Y)
cat("variance of the phenotype: ", var(phenores$Y), "\n")

snpname = dat$snp[,1]
anno <- cbind(dat$chr, dat$pos, dat$pos)

outname <- "/home/simingz/causalTWAS/simulations/simulation_susieI_20201001/chr22_plots/chr22snpRandom"
# GWAA(dat$G, matrix(rnorm(N, sd = sqrt(var(phenores$Y))), ncol =1), snpname = snpname, anno = anno, outname , family = gaussian, ncore = 5, nSplits = 30, compress = F)
n <- read.table(outname, header =T, comment.char = "")
plot(n$BEGIN, -log10(n$PVALUE), main = "Random", xlab = "position", ylab = "-log10(p)")
grid()
```

## genetic component of chromosome 2 as phenotype

Not all chromosome give small p values. chr2 is the one with the smallest p values.
```{r chrfixed, fig.height=4, fig.width=8}
chrom.p <- 2
r <- do.call(cbind, lapply(1:22, function(x) phenores$batch[[x]]$Y.g))
cat("variance of the genetic component: ", var(r[,chrom.p]))

outname <- "/home/simingz/causalTWAS/simulations/simulation_susieI_20201001/chr22_plots/chr22snpVschr2fixed"
# GWAA(dat$G, matrix(r[,chrom.p], ncol =1), snpname = snpname, anno = anno, outname, family = gaussian, ncore = 5, nSplits = 30, compress = F)
n <- read.table(outname, header =T, comment.char = "", stringsAsFactors = F)
plot(n$BEGIN, -log10(n$PVALUE), main = "chr2 genetic", xlab = "position", ylab = "-log10(p)")
grid()
```

## genotype of causal snps on chromosome 2 as phenotype

genotype of causal snps on chromosome 2 as phenotype. There are `r length(phenores$batch[[2]]$param$idx.cSNP)` causal SNPs. Not all of them significantly associated with SNP genotype on chr22, but quite a few. below are examples (only a subset of SNPs on chr22 are selected to run regression analysis with chr2 causal SNP genotypes.) 

```{r geno, fig.height=4, fig.width=8, message=F}
chrom <- 22
load(paste0("~/causalTWAS/ukbiobank/ukb_chr", chrom ,"_s40000.FBM.Rd"))
geno <- dat$G[, match(n[n$PVALUE < 1e-40, "MARKER_ID"], dat$snp[,1])]
load(paste0("~/causalTWAS/ukbiobank/ukb_chr", chrom ,"_s40000.unscaled.FBM.Rd"))
geno.unscaled <-  dat$G[, match(n[n$PVALUE < 1e-40, "MARKER_ID"], dat$snp[,1])]

snpname = dat$snp[match(n[n$PVALUE < 1e-40, "MARKER_ID"], dat$snp[,1]),1]
anno <- cbind(dat$chr, dat$pos, dat$pos)[match(n[n$PVALUE < 1e-40, "MARKER_ID"], dat$snp[,1]), ]

load(paste0("~/causalTWAS/ukbiobank/ukb_chr", chrom.p ,"_s40000.FBM.Rd"))
geno.c <- dat$G[, phenores$batch[[chrom.p]]$param$idx.cSNP]
geno.c.names <- dat$snp[phenores$batch[[chrom.p]]$param$idx.cSNP,]
load(paste0("~/causalTWAS/ukbiobank/ukb_chr", chrom.p ,"_s40000.unscaled.FBM.Rd"))
geno.c.unscaled <- dat$G[, phenores$batch[[chrom.p]]$param$idx.cSNP]


ac <- ac <- apply(round(geno.unscaled), 2, table)
maf <- apply(cbind( (ac[1,] * 2  + ac[2,]) / N/2, 1- (ac[1,] * 2  + ac[2,]) /N/2), 1, min)
  
geno.c.select <- c(1, 23, 45, 65)

for (i in geno.c.select){
  outname <- paste0("/home/simingz/causalTWAS/simulations/simulation_susieI_20201001/chr22_plots/chr22snpVschr2causal", i)
  #GWAA(geno, matrix(geno.c[,i], ncol =1), snpname = snpname, anno = anno, outname, family = gaussian, ncore = 5, nSplits = 10, compress = F)
  rsq_max <- ld_max(geno.c.unscaled[,i, drop =F], geno.unscaled, stats = "R.squared")

  cat("max LD: ", rsq_max)
  m <- read.table(outname, header =T, comment.char = "")
  rbPal <- colorRampPalette(c('red','blue'))
  plot(anno[,2], -log10(m[m$X.CHROM ==22, ]$PVALUE),  main = paste0("chr", chrom.p, ":", geno.c.names[i]), xlab = "position", ylab = "-log10(p)", col =rbPal(11)[as.numeric(cut(maf,breaks = seq(0,0.5, 0.05)))], pch =20)
  legend("topleft",title="MAF",legend=seq(0,0.5, 0.05),col =rbPal(11),pch=20, cex = 0.5)
  grid()
}
```

# Parameter estimation results

Results: Each row shows parameter estimation results from 5 simulation runs with similar settings (i.e. pi1 and PVE for genes and SNPs). each row has two plots, one for gene pi1 estimation, one for enrichment (gene pi1/snp pi1). Results from each run were represented by one dot, dots with the same color come from the same run. horizontal dash lines: simulation truth,  `susietruth`, the truth in selected regions that were used to run susie iteractively (susieI).


```{r param functions}
show_param <- function(phenofs, susieIfs, susieIfs2){
  pars <- do.call(rbind, lapply(phenofs, function(x) {load(x); 
    c(phenores$param$pve.gene.truth,
      phenores$param$pve.snp.truth,
      length(phenores$batch[[1]]$param$idx.cgene)/phenores$batch[[1]]$param$J,
      length(phenores$batch[[1]]$param$idx.cSNP)/phenores$batch[[1]]$param$M)}))
 
  colnames(pars) <- c("PVE.gene_truth", "PVE.SNP_truth", "pi1.gene_truth", "pi1.SNP_truth")
    
  param.s <- do.call(rbind, lapply(susieIfs, function(x) {load(x); c(tail(prior.gene_rec[prior.gene_rec!=0], 1), tail(prior.SNP_rec[prior.SNP_rec!=0],1))}))
  
  param.s.truth <- do.call(rbind, lapply(susieIfs2, function(x) {
    a <- fread(x, header = T);
    c(nrow(a[a$ifcausal == 1 & a$type == "gene" ])/ nrow(a[a$type == "gene"]),
      nrow(a[a$ifcausal == 1 & a$type == "SNP"])/ nrow(a[a$type == "SNP"]))
    }))
  
  pars.s <- cbind(param.s.truth, param.s)[, c(1,3,2,4)]
  colnames(pars.s) <-  paste(rep(c("pi1.gene_", "pi1.SNP_"), each = 2), c("susietruth", "susieI"), sep = "")
  
  df <- cbind(tags, format(pars, digits = 4), format(pars.s, digits =4))
  rownames(df) <- NULL
  return(df)
  
  # df %>% 
  # kable("html", escape = F) %>%
  # kable_styling("striped", full_width = F) %>%
  #  row_spec(c(1:5, 11:15), background = "#FEF3B9") %>%
  # scroll_box(width = "100%", height = "600px", fixed_thead = T)
}

plot_param <- function(df, ...){
  df <- apply(df[ , 2:ncol(df)], 2, function(x) as.numeric(x))
  st <-  cbind(df[,"pi1.gene_susietruth"], 1:nrow(df), 2 + 1:nrow(df)/nrow(df)/3)
  s <- cbind(df[,"pi1.gene_susieI"], 1:nrow(df), 3 + 1:nrow(df)/nrow(df)/3)
  t <- df[1,"pi1.gene_truth"]
  dfp <- rbind(st,s)
  plot(dfp[,3], dfp[,1], col = dfp[,2], pch = 19, ylab = "gene pi1", xaxt = "n", xlab="", xlim = c(0.8, 3.5), frame.plot=FALSE, ylim = c(0, max(dfp[,1],t) *1.05), ...)
  axis(side=1, at=1:2, labels = FALSE, tick = F)
  text(x=2:3, 0, labels = c( "susieI_truth", "susieI"), xpd = T, pos =1)
  abline(h=t, lty = 2, col= "salmon", lwd=1.5)
  grid()
  
  st <-  cbind(df[,"pi1.gene_susietruth"]/df[,"pi1.SNP_susietruth"], 1:nrow(df), 2 + 1:nrow(df)/nrow(df)/3)
  s <- cbind(df[,"pi1.gene_susieI"]/df[,"pi1.SNP_susieI"], 1:nrow(df), 3 + 1:nrow(df)/nrow(df)/3)
  t <- df[1,"pi1.gene_truth"]/df[1,"pi1.SNP_truth"]
  dfp <- rbind(st,s)
  plot(dfp[,3], dfp[,1], col = dfp[,2], pch = 19, ylab = "Enrichment (gene/snp)", xaxt = "n", xlab="", xlim = c(0.8, 3.5),frame.plot=FALSE, ylim = c(0, min(max(dfp[,1],t) *1.05, 150)))
  axis(side=1, at=1:2, labels = FALSE, tick = F)
  text(x=2:3, 0, labels = c("susieI_truth", "susieI"), xpd = T, pos =1)
  abline(h= t, lty = 2, col= "darkgreen", lwd=1.5)
  grid()
}

gpip_dist <- function(susiefs, ...){
  dflist <- list()
  for (f in susiefs){
    dflist[[f]] <- read.table(f, header =T , stringsAsFactors = F)
  }
  df <- do.call(rbind, dflist)
  hist(df[df$type == "gene", "susie_pip"], xlab = "gene susie PIP", 
       breaks = 50, ylim = c(0,20), xlim=c(0,1), col = "salmon", ...)
}
```



## Regions: all (1)

* Susie run parameters: `L=2`. initialize with `null_weight = 0` and update `null_weight` based on last iteration results. We initialize with prior for genes and SNPs as uniform. gene.pve ~ 0.1, snp.pve ~ 0.5. 

```{r param1, fig.height=6, fig.width=6}
tags <- paste(rep(1:2, each = 5), 1:5, sep = "-")
phenofs <- paste0(simdatadir, "simu_20201001-", tags, "-pheno.Rd")
susieIfs <- paste0(outputdir, "20201001-", tags, ".config1.susieIres.Rd")
susieIfs2 <- paste0(outputdir, "20201001-", tags, ".config1.susieI.txt")

df <- show_param(phenofs, susieIfs, susieIfs2)
par(mfrow = c(2,2))
plot_param(df[1:5,], main = "low power")
plot_param(df[6:10,], main = "high power")
```

## Regions: causal

* Susie run parameters: `L=2`. initialize with `null_weight = 0` and update `null_weight` based on last iteration results. We initialize with prior for genes and SNPs as uniform. gene.pve ~ 0.1, snp.pve ~ 0.5. 

```{r param5, fig.height=6, fig.width=6}
tags <- paste(rep(1:2, each = 5), 1:5, sep = "-")
phenofs <- paste0(simdatadir, "simu_20201001-", tags, "-pheno.Rd")
susieIfs <- paste0(outputdir, "20201001-", tags, ".config5.susieIres.Rd")
susieIfs2 <- paste0(outputdir, "20201001-", tags, ".config5.susieI.txt")

df <- show_param(phenofs, susieIfs, susieIfs2)
par(mfrow = c(2,2))
plot_param(df[1:5,], main = "low power")
plot_param(df[6:10,], main = "high power")
```

