---
title: "Test a simpler version of veb-boost performance on ukbiobank data"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r plot functions}
library(mr.ash.alpha)
source("~/causalTWAS/causal-TWAS/code/fit_mr.ash.R")

summary_mr.ash <- function(fit){
  cat("pi1 = ", 1-fit$pi[[1]], "\n")
  pve <- get_pve(fit)
  cat("pve = ", pve, "\n")
}

plot_beta <- function(beta,beta.pm, ...){
  plot( beta, pch=19, col ="darkgreen", ...)
  points(beta.pm, pch =19, col = "red")
  legend("topright", legend=c("true beta", "posterior mean"),
       col=c("darkgreen", "red"), pch=19)
}

plot_pip <- function(indi, pip, pipcut = 0.1) {
  x <- barplot(indi, main = "PIP")
  points(x, pip, pch=19, col="red")
  pred <- which(pip > pipcut)
  real <- which(indi==1)
  tp <- length(intersect(pred, real))/length(pred)
  fp <- 1 - tp
  fn <- length(setdiff(real, pred))/length(real)
  cat("false positive rate: ", fp, "\n")
  cat("false negative rate: ", fn, "\n")
}

summary_mr.ash2 <- function(g.fit, s.fit, phenores, pipcut = 0.1){
  e.b <- rep(0, length(g.fit$beta))
  e.b[phenores$param$idx.cgene] <- phenores$param$e.beta
  cat("gene expression effect: \n")
  summary_mr.ash(g.fit)
  plot_beta(e.b, g.fit$beta)
  indi <- rep(0, length(g.fit$beta))
  indi[phenores$param$idx.cgene] <- 1
  pip <- get_pip(g.fit)
  plot_pip(indi, pip, pipcut = pipcut)
  
  s.b <- rep(0, length(s.fit$beta))
  s.b[phenores$param$idx.cSNP] <- phenores$param$s.theta
  cat("snp effect: \n")
  summary_mr.ash(s.fit)
  plot_beta(s.b, s.fit$beta)
}

```

```{r output dir}
outputdir <- "~/causalTWAS/simulations/simulation_ashtest_20200503/"
```

## Simulation 1
True parameters:
```{r para 1}
load(paste0(outputdir, "mr.ash2_20200503-1-pheno.Rd"))
readLines(paste0(outputdir, "param-20200503-1.R"))
```

MR.ASH results (start from gene):
```{r sim 1 gene}
load(paste0(outputdir, "mr.ash2_20200503-1-mr.ash2.expr-res.Rd"))
g.fit <- mr.ash2.fit$fit1
s.fit <-  mr.ash2.fit$fit2
summary_mr.ash2(g.fit, s.fit, phenores)
```

MR.ASH results (start from SNP):
```{r sim 1 snp}
load(paste0(outputdir, "mr.ash2_20200503-1-mr.ash2.snp-res.Rd"))
g.fit <- mr.ash2.fit$fit2
s.fit <-  mr.ash2.fit$fit1
summary_mr.ash2(g.fit, s.fit, phenores)
```

## Simulation 2

True parameters:
```{r para 2}
load(paste0(outputdir, "mr.ash2_20200503-2-pheno.Rd"))
readLines(paste0(outputdir, "param-20200503-2.R"))
```

MR.ASH results (start from gene):
```{r sim 2 gene}
load(paste0(outputdir, "mr.ash2_20200503-2-mr.ash2.expr-res.Rd"))
g.fit <- mr.ash2.fit$fit1
s.fit <-  mr.ash2.fit$fit2
summary_mr.ash2(g.fit, s.fit, phenores)
```

MR.ASH results (start from SNP):
```{r sim 2 snp}
load(paste0(outputdir, "mr.ash2_20200503-2-mr.ash2.snp-res.Rd"))
g.fit <- mr.ash2.fit$fit2
s.fit <-  mr.ash2.fit$fit1
summary_mr.ash2(g.fit, s.fit, phenores)
```

## Simulation 3

True parameters:
```{r para 3}
load(paste0(outputdir, "mr.ash2_20200503-3-pheno.Rd"))
readLines(paste0(outputdir, "param-20200503-3.R"))
```

MR.ASH results (start from gene):
```{r sim 3 gene}
load(paste0(outputdir, "mr.ash2_20200503-3-mr.ash2.expr-res.Rd"))
g.fit <- mr.ash2.fit$fit1
s.fit <-  mr.ash2.fit$fit2
summary_mr.ash2(g.fit, s.fit, phenores)
```

MR.ASH results (start from SNP):
```{r sim 3 snp}
load(paste0(outputdir, "mr.ash2_20200503-3-mr.ash2.snp-res.Rd"))
g.fit <- mr.ash2.fit$fit2
s.fit <-  mr.ash2.fit$fit1
summary_mr.ash2(g.fit, s.fit, phenores)
```

## Simulation 4

True parameters:
```{r para 4}
load(paste0(outputdir, "mr.ash2_20200503-4-pheno.Rd"))
readLines(paste0(outputdir, "param-20200503-4.R"))
```

MR.ASH results (start from gene):
```{r sim 4 gene}
load(paste0(outputdir, "mr.ash2_20200503-4-mr.ash2.expr-res.Rd"))
g.fit <- mr.ash2.fit$fit1
s.fit <-  mr.ash2.fit$fit2
summary_mr.ash2(g.fit, s.fit, phenores)
```

MR.ASH results (start from SNP):
```{r sim 4 snp}
load(paste0(outputdir, "mr.ash2_20200503-4-mr.ash2.snp-res.Rd"))
g.fit <- mr.ash2.fit$fit2
s.fit <-  mr.ash2.fit$fit1
summary_mr.ash2(g.fit, s.fit, phenores)
```
