---
title: "Simulations to test ctwas summary stats version, 113k samples, a few new setings required"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,  message= F, warning = F)
```

```{r libs}
library(ctwas)
library(data.table)
suppressMessages({library(plotly)})
library(tidyr)
library(plyr)
library(stringr)
source("~/causalTWAS/causal-TWAS/analysis/summarize_ctwas_plots.R")
source('~/causalTWAS/causal-TWAS/analysis/summarize_twas-coloc_plots.R')
source('~/causalTWAS/causal-TWAS/analysis/summarize_focus_plots.R')
source('~/causalTWAS/causal-TWAS/analysis/summarize_smr_plots.R')
source('~/causalTWAS/causal-TWAS/analysis/summarize_mrjti_plots.R')
source('~/causalTWAS/causal-TWAS/code/qqplot.R')
```

```{r files}
pgenfn = "/home/simingz/causalTWAS/ukbiobank/ukb_pgen_s200.113/ukb-s200.113_pgenfs.txt"
ld_pgenfn = "/home/simingz/causalTWAS/ukbiobank/ukb_pgen_s200.113/ukb-s200.113.2_pgenfs.txt"
outputdir = "/home/simingz/causalTWAS/simulations/simulation_ctwas_rss_20230321/" # /
comparedir = "/home/simingz/causalTWAS/simulations/simulation_ctwas_rss_20230321_compare/"
runtag = "ukb-s200.113-adi"

simutags = paste(rep(1:9, each = length(1:5)), 1:5, sep = "-")

pgenfs <- read.table(pgenfn, header = F, stringsAsFactors = F)[,1]
pvarfs <- sapply(pgenfs, prep_pvar, outputdir = outputdir)

ld_pgenfs <- read.table(ld_pgenfn, header = F, stringsAsFactors = F)[,1]
ld_pvarfs <- sapply(ld_pgenfs, prep_pvar, outputdir = outputdir)

pgens <- lapply(1:length(pgenfs), function(x) prep_pgen(pgenf = pgenfs[x],pvarf = pvarfs[x]))
```

# Analysis description

```{r parameters}
n.ori <- 200000 # number of samples
n <- pgenlibr::GetRawSampleCt(pgens[[1]])
p <- sum(unlist(lapply(pgens, pgenlibr::GetVariantCt))) # number of SNPs
J <- 8021 # number of genes
```

## Data
The same as 20210416. We tried a few additional settings with small SNP heritablity (0.2) or large gene heritablity (0.5) following the suggestions by reviewers. 
 
## Analysis

### ctwas

1. Get z scores for gene expression. 
We used expression models and LD reference to get z scores for gene expression.

2. Run ctwas_rss
`ctwas_rss` algorithm first runs on all regions to get rough estimate for gene and SNP prior. Then run on small regions (having small probablities of having > 1 causal signals based on rough estimates) to get more accurate estimate. To lower computational burden, we downsampled SNPs (0.1) to estimate parameters. With the estimated parameters, we then run susie for all regions using both genes and downsampled SNPs with specified $L$. After this, for regions with strong gene signals, we rerun susie with full SNPs using specified $L$. 

Configurations

`ld_regions ='EUR'`, We used LDetect to define regions. To match UKbiobank data, we use the 'EUR' population

`thin = 0.1`, downsampled SNPs to 1/10 for parameter estimation step

`niter1 =3`, run `niter1 =3` iterations first to get some rough parameter estimates.

`prob_single = 0.8`, the probability of a region having at most 1 singal has to be at least 0.8 to be selected for the parameter estimation step. This probability is obtained by using the PIPs from the first few iterations.

`niter2 = 30`, run `niter2 = 30` for parameter estimation step

`group_prior = NULL`, the initiating prior parameters we used for running susie for each region is uniform prior for genes and SNPs.

`group_prior_var = NULL`, the initiating prior variance parameters we used for running susie for each region follows susie_rss's default (50).

`max_SNP_region = 5000`, the maximum number of SNPs for re-running susie on strong gene signal regions is 5000. 

We have two configurations for `L`: config 1 the last step of ctwas was running susie with `L=5` in regions with big gene PIPs, config 2 the last step of ctwas was running susie with `L=1` in regions with big gene PIPs.

# `ctwas` results

Results: Each row shows parameter estimation results from >5 simulation runs with similar settings (i.e. pi1 and PVE for genes and SNPs). Results from each run were represented by one dot, dots with the same color come from the same run. `truth`: the true parameters,  `selected_truth`: the truth in selected regions that were used to estimate parameters, `ctwas`: ctwas estimated parameters (using summary statistics as input). 

Note, config 1 was not run correctly. Need to rerun with `harmonize_z = F` in `ctwas_rss`.


```{r func}
plot_par <- function(configtag, runtag, simutags){
  source(paste0(outputdir, "config", configtag, ".R"))
  phenofs <- paste0(outputdir, runtag, "_simu", simutags, "-pheno.Rd")
  susieIfs <- paste0(outputdir, runtag, "_simu", simutags, "_config", configtag, ".s2.susieIrssres.Rd")
  susieIfs2 <- paste0(outputdir, runtag, "_simu",simutags, "_config", configtag,".s2.susieIrss.txt")

  mtx <- show_param(phenofs, susieIfs, susieIfs2, thin = thin)
  par(mfrow=c(1,3))
  cat("simulations ", paste(simutags, sep=",") , ": ")
  cat("mean gene PVE:", mean(mtx[, "PVE.gene_truth"]), ",", "mean SNP PVE:", mean(mtx[, "PVE.SNP_truth"]), "\n")
  plot_param(mtx)
}

plot_PIP <- function(configtag, runtag,  simutags){
   phenofs <- paste0(outputdir, runtag, "_simu", simutags, "-pheno.Rd")
   susieIfs <- paste0(outputdir, runtag, "_simu",simutags, "_config", configtag,".susieIrss.txt")

   f1 <- caliPIP_plot(phenofs, susieIfs)
   f2 <- ncausal_plot(phenofs, susieIfs) 
   gridExtra::grid.arrange(f1, f2, ncol =2)
}

plot_fusion_coloc <- function(configtag, runtag,  simutags){
    phenofs <- paste0(outputdir, runtag, "_simu", simutags, "-pheno.Rd")
    fusioncolocfs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.coloc.result")
    
    f1 <- caliFUSIONp_plot(phenofs, fusioncolocfs)
    f2 <- ncausalFUSIONp_plot(phenofs, fusioncolocfs)
    f3 <- caliFUSIONbon_plot(phenofs, fusioncolocfs)
    f4 <- ncausalFUSIONbon_plot(phenofs, fusioncolocfs)
    f5 <- caliPP4_plot(phenofs, fusioncolocfs, twas.p = 0.05/J)
    f6 <- ncausalPP4_plot(phenofs, fusioncolocfs, twas.p = 0.05/J)
    gridExtra::grid.arrange(f1, f2, ncol=2)
    gridExtra::grid.arrange(f3, f4, ncol=2)
    gridExtra::grid.arrange(f5, f6, ncol=2)
}
    
plot_focus <- function(configtag, runtag,  simutags){
    phenofs <- paste0(outputdir, runtag, "_simu", simutags, "-pheno.Rd")
    focusfs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.focus.tsv")
    
    f1 <- califocusPIP_plot(phenofs, focusfs)
    f2 <- ncausalfocusPIP_plot(phenofs, focusfs)
    gridExtra::grid.arrange(f1, f2, ncol=2)
}

plot_smr <- function(configtag, runtag,  simutags){
    phenofs <- paste0(outputdir, runtag, "_simu", simutags, "-pheno.Rd")
    smrfs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.smr")
    
    f1 <- caliSMRp_plot(phenofs, smrfs)
    f2 <- ncausalSMRp_plot(phenofs, smrfs)
    gridExtra::grid.arrange(f1, f2, ncol=2)
}

plot_mrjti <- function(configtag, runtag,  simutags){
    phenofs <- paste0(outputdir, runtag, "_simu", simutags, "-pheno.Rd")
    mrfs <- paste0(comparedir, runtag, "_simu", simutags, ".Adipose_Subcutaneous.mrjti.result")
    
    f1 <- caliMR_plot(phenofs, mrfs)
    f2 <- ncausalMR_plot(phenofs, mrfs)
    gridExtra::grid.arrange(f1, f2, ncol=2)
}
```


```{r config-1-2, fig.height=3, fig.width=8, message=F}

runtag = "ukb-s200.113-adi"

# has not rerun last step for simu 1-x. 
# simutags <- paste(1, c(1:24, 26:54, 56:70), sep = "-")
# configtag <- 1 
# plot_par(configtag, runtag, simutags)
# plot_PIP(configtag, runtag, simutags)
# plot_par(1, runtag, simutags)
# print("When running with L= 5 in final step:")
# plot_PIP(1, runtag, simutags)
# print("When running with L= 1 in final step:")
# plot_PIP(2, runtag, simutags)
#plot_fusion_coloc(configtag, runtag, simutags)
# plot_focus(configtag, runtag, simutags)
#print("FOCUS result is null")
#plot_smr(configtag, runtag, simutags)

simutags <- paste(2, 1:30, sep = "-")
plot_par(1, runtag, simutags)
print("When running with L= 5 in final step:")
plot_PIP(1, runtag, simutags)
print("When running with L= 1 in final step:")
plot_PIP(2, runtag, simutags)

simutags <- paste(6, c(1:20), sep = "-")
plot_par(1, runtag, simutags)
print("When running with L= 5 in final step:")
plot_PIP(1, runtag, simutags)
print("When running with L= 1 in final step:")
plot_PIP(2, runtag, simutags)

simutags <- paste(3, 1:10, sep = "-")
plot_par(1, runtag, simutags)
print("When running with L= 5 in final step:")
plot_PIP(1, runtag, simutags)
print("When running with L= 1 in final step:")
plot_PIP(2, runtag, simutags)

simutags <- paste(4, 1:10, sep = "-")
plot_par(1, runtag, simutags)
print("When running with L= 5 in final step:")
plot_PIP(1, runtag, simutags)
print("When running with L= 1 in final step:")
plot_PIP(2, runtag, simutags)

simutags <- paste(5, c(1:10), sep = "-")
plot_par(1, runtag, simutags)
print("When running with L= 5 in final step:")
plot_PIP(1, runtag, simutags)
print("When running with L= 1 in final step:")
plot_PIP(2, runtag, simutags)
```

