---
title: "Simulation-WTCCC"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
Use WTCCC data obtained from Peter on CRI. 

## Simulate cis-expression 
We simulate cis- expression based on the following model:

$$ \tilde{X} = G\alpha + \xi$$
For each gene, we sample $L$ eQTLs for this gene, for eQTL $k$, we have
$$ \alpha_k \sim N(0,\sigma_\alpha^2) $$
$$ \xi \sim N(0,1) $$
Then we have the heritability of the gene: 

$$\begin{aligned}
h^2_{eQTL} &= \frac{var(G\alpha)}{var(G\alpha)+var(\xi)} \\
&= \frac{\Sigma_kvar(G_k) \alpha_k^2}{\Sigma_kvar(G_k) \alpha_k^2 + var(\xi)}\\
&=  \frac{\Sigma_k\alpha_k^2}{\Sigma_k\alpha_k^2 + var(\xi)}\\
&\approx \frac{\Sigma_k\sigma_\alpha^2}{\Sigma_k\sigma_\alpha^2 + var(\xi)} \\
&=\frac{K\sigma_\alpha^2}{1+K\sigma_\alpha^2}
\end{aligned}$$

Here, we use scaled genotype data, so $var(G)=1$. We also have $\alpha^2_k \approx E(\alpha_k^2)=var(\alpha_k^2)=\sigma_\alpha^2$. Usually, $K$ ranges from 1 to 5, let $\sigma_\alpha = 0.3$, then $h^2_{eQTL}$ ranges from 0.08 to 0.31, this is consistent with gene cis-heritability in reality.

## Simulate quantitative phenotype

$$\begin{align}
PVE_{SNP} = &\frac{var(G\theta)}{var(G\theta) + var(\tilde{X}\gamma) + \sigma_e^2}\\ 
\approx &\frac{M\sigma_\theta^2}{M\sigma_\theta^2 + \Sigma_jvar(\tilde{X_j})\gamma_j^2+ \sigma_e^2}\\ 
\approx &\frac{M\sigma_\theta^2}{M\sigma_\theta^2 + Jvar(\tilde{X})\sigma_\gamma^2+ \sigma_e^2}\\   
\end{align}$$

$$
PVE_{expr} \approx \frac{Jvar(\tilde{X})\sigma_\gamma^2}{M\sigma_\theta^2 + Jvar(\tilde{X})\sigma_\gamma^2+ \sigma_e^2}$$

Here, $var(\tilde{X})$ is the cis-heratbility of gene expression. $M$ and $J$ are number of causal SNP and gene respectively. 

Thus in order to get desired $PVE_{SNP}$ and $PVE_expr$, we set $\sigma_\theta^2$ and $\sigma_\gamma^2$ based on the following formula:

$$\sigma_\theta^2 = \frac{PVE_{SNP}}{M(1-PVE_{SNP} - PVE_{expr})}$$
$$\sigma_\gamma^2 = \frac{PVE_{expr}}{Jvar(\tilde{X})(1-PVE_{SNP} - PVE_{expr})}$$

## Simulation 1: $M$ = No. of all SNPs, $J$ = No. all genes
We simulate quantitative phenotype under several scenarios. First we make all genes with imputed gene expression as causal genes and all SNPs as causal SNPs. This matches our polygenic version of the model. 

### 1.1 $PVE_{SNP}=0.1$, $PVE_{expr}=0.1$
We set $\sigma_e=1$. 

### 1.2 $PVE_{SNP}=0.1$, $PVE_{expr}=0.4$
We set $\sigma_e=1$. 

## Simulation 2: $M$ = No. of all SNPs, $J$ = ~10% all genes (J=2000)

### 2.1 $PVE_{SNP}=0.1$, $PVE_{expr}=0.1$
We set $\sigma_e=1$. 

### 2.2 $PVE_{SNP}=0.1$, $PVE_{expr}=0.4$
We set $\sigma_e=1$. 

## Simulation 3: $M$ = ~ 10% all SNPs (M=50000), $J$ = No. all genes

### 3.1 $PVE_{SNP}=0.1$, $PVE_{expr}=0.1$ 
We set $\sigma_e=1$. 

### 3.2 $PVE_{SNP}=0.1$, $PVE_{expr}=0.4$
We set $\sigma_e=1$. 

## Simulation 4: $M$ = ~ 10% all SNPs (M=50000), $J$ = ~10% all genes (J=2000)

### 4.1 $PVE_{SNP}=0.1$, $PVE_{expr}=0.1$ 
We set $\sigma_e=1$. 

### 4.2 $PVE_{SNP}=0.1$, $PVE_{expr}=0.4$
We set $\sigma_e=1$. 




