---
title: "Simulation-WTCCC-2"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
Use WTCCC data obtained from Peter on CRI. 

The difference between this analysis and [previous run](simulation-WTCCC.html): used 500 samples from `cad.RData` from WTCCC to simulate expression, then train expression model using `cv.glmnet`. We impute expression in using `bd.RData` from WTCCC. 

Simulation of phenotype: use the same weights as used in simulate expression using the `'cad.RData` dataset, simulate phenotype according to:

$$Y = G\alpha\gamma + G\theta + \epsilon$$. For $\alpha$, we used the same as used in simulating expression (Thus, not $\tilde{X}$). $\gamma \sim N(0,\sigma_\gamma^2), \theta \sim N(0, \sigma_\theta^2), \epsilon \sim N (0,1)$.

The genotype data and other settings are the same.

## Simulate expression 
We simulate expression based on the following model, using 500 samples from cad in WTCCC:

$$ X = G\alpha + \xi$$
For each gene, we sample $L$ eQTLs for this gene, for eQTL $k$, we have
$$ \alpha_k \sim N(0,\sigma_\alpha^2) $$
$$ \xi \sim N(0,1) $$
Then we have the heritability of the gene: 

$$\begin{aligned}
h^2_{eQTL} &= \frac{var(G\alpha)}{var(G\alpha)+var(\xi)} \\
&= \frac{\Sigma_kvar(G_k) \alpha_k^2}{\Sigma_kvar(G_k) \alpha_k^2 + var(\xi)}\\
&=  \frac{\Sigma_k\alpha_k^2}{\Sigma_k\alpha_k^2 + var(\xi)}\\
&\approx \frac{\Sigma_k\sigma_\alpha^2}{\Sigma_k\sigma_\alpha^2 + var(\xi)} \\
&=\frac{K\sigma_\alpha^2}{1+K\sigma_\alpha^2}
\end{aligned}$$

Here, we use scaled genotype data, so $var(G)=1$. We also have $\alpha^2_k \approx E(\alpha_k^2)=var(\alpha_k^2)=\sigma_\alpha^2$. Usually, $K$ ranges from 1 to 5, let $\sigma_\alpha = 0.3$, then $h^2_{eQTL}$ ranges from 0.08 to 0.31, this is consistent with gene cis-heritability in reality.

## Simulate quantitative phenotype
We simulate phenotype using bd data from WTCCC following
$$Y = G\alpha\gamma + G\theta + \epsilon$$. For $\alpha$, we used the same as used in simulating expression. $\gamma \sim N(0,\sigma_\gamma^2), \theta \sim N(0, \sigma_\theta^2), \epsilon \sim N (0,1)$.

$$\begin{align}
PVE_{SNP} = &\frac{var(G\theta)}{var(G\theta) + var(\tilde{X}\gamma) + \sigma_e^2}\\ 
\approx &\frac{M\sigma_\theta^2}{M\sigma_\theta^2 + \Sigma_jvar(\tilde{X_j})\gamma_j^2+ \sigma_e^2}\\ 
\approx &\frac{M\sigma_\theta^2}{M\sigma_\theta^2 + Jvar(\tilde{X})\sigma_\gamma^2+ \sigma_e^2}\\   
\end{align}$$

$$
PVE_{expr} \approx \frac{Jvar(\tilde{X})\sigma_\gamma^2}{M\sigma_\theta^2 + Jvar(\tilde{X})\sigma_\gamma^2+ \sigma_e^2}$$

Here, $var(\tilde{X})$ is the cis-heratbility of gene expression. $M$ and $J$ are numbers of causal SNP and gene respectively. 

Thus in order to get desired $PVE_{SNP}$ and $PVE_{expr}$, we set $\sigma_\theta^2$ and $\sigma_\gamma^2$ based on the following formula:

$$\sigma_\theta^2 = \frac{PVE_{SNP}}{M(1-PVE_{SNP} - PVE_{expr})}$$
$$\sigma_\gamma^2 = \frac{PVE_{expr}}{Jvar(\tilde{X})(1-PVE_{SNP} - PVE_{expr})}$$

```{r}
simudir <- "/home/simingz/causalTWAS/simulations/simulation_WTCCC_20191127/"
load("/home/simingz/causalTWAS/WTCCC/bd.toy.RData")
Gvar <- mean(apply(X,2,var))
show_res <- function(simutag){
  load(Sys.glob(paste0(simudir,simutag,"*phenotype.Rd")))
  outdf1 <- data.frame(truth = c(sigma_gamma^2*J.c, sigma_theta^2*M.c/Gvar , 1))
  row.names(outdf1) <- c("sigma_gamma^2","sigma_theta^2","sigma_e^2")
  res <- readLines(paste0(simudir, simutag, "_VC/output/", simutag, ".log.txt"))
  outdf1$est <- as.numeric(strsplit(res[24], "  ")[[1]][2:4])
  outdf1$est.se <- as.numeric(strsplit(res[25], "  ")[[1]][2:4])
  print(outdf1)
  outdf2 <- data.frame("est"= c(as.numeric(strsplit(res[20], "  ")[[1]][2:3]), as.numeric(strsplit(res[22], "=")[[1]][2])))
  row.names(outdf2) <- c("PVE.expr","PVE.snp","PVE.total")
  outdf2$est.se <- c(as.numeric(strsplit(res[21], "  ")[[1]][2:3]), as.numeric(strsplit(res[23], "=")[[1]][2]))
  print(outdf2)
}
```

## Simulation 1: $M$ = No. of all SNPs, $J$ = No. all genes
We simulate quantitative phenotype under several scenarios. First we make all genes with imputed gene expression as causal genes and all SNPs as causal SNPs. This matches our polygenic version of the model. 

### 1.1 $PVE_{SNP} \approx 0.1$, $PVE_{expr} \approx 0.1$
```{r 1.1}
show_res("S1.1")
```

### 1.2 $PVE_{SNP}\approx 0.4$, $PVE_{expr}\approx 0.1$
```{r 1.2}
show_res("S1.2")
```

## Simulation 2: $M$ = No. of all SNPs, $J$ = ~10% all genes (J=2000)

### 2.1 $PVE_{SNP} \approx 0.1$, $PVE_{expr} \approx 0.1$
```{r 2.1}
show_res("S2.1")
```

### 2.2 $PVE_{SNP} \approx 0.4$, $PVE_{expr} \approx 0.1$
```{r 2.2}
show_res("S2.2")
```
## Simulation 3: $M$ = ~ 1% all SNPs (M=5000), $J$ = No. all genes

### 3.1 $PVE_{SNP} \approx 0.1$, $PVE_{expr} \approx 0.1$ 
```{r 3.1}
show_res("S3.1")
```

### 3.2 $PVE_{SNP} \approx 0.4$, $PVE_{expr} \approx 0.1$
```{r 3.2}
show_res("S3.2")
```

## Discussion
In this analysis, we see the PVE estimation for expression is much under estimated, this is because many of the gene expression training model has low $R^2$, especially for low heritability ones. For a polygenic model, because we are estimating the total herability explained by expression, the ones with low $R^2$ will affect this estimation. We are moving to model with a sparse prior for expression.




