---
title: "Test susieI using whole genome, s80.45 samples"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,  message= F, warning = F)
```
```{r libs}
library(mr.ash.alpha)
library(data.table)
suppressMessages({library(plotly)})
library(tidyr)
library(plyr)
library(stringr)
source("analysis/summarize_twas_plots.R")
```

# Analysis description

```{r parameters}
n.ori <- 80000 # number of samples
n <-  45542
p <- 6656321 # number of SNPs
J <- 8021 # number of genes
```

The genotype data we used is from UKB biobank, randomly selecting `r as.character(n.ori)` samples. We then filtered samples based on relatedness, ethics and other qc metrics, that ended up with n = `r as.character(n)` samples.  We use SNP genotype data from chr 1 to chr 22 combined from UKB. There total = `r as.character(p)` SNPs.

Our analysis consists of the following steps: 

1. Build expression predictors using another expression-genotype dataset. 

The one we used in this analysis is GTEx Adipose tissue v7 dataset. This dataset contains ~ 380 samples. [FUSION/TWAS](http://gusevlab.org/projects/fusion/) were used to train expression model and we used their lasso results. SNPs included in eQTL anlaysis are restricted to cis-locus 500kb on either side of the gene boundary. eQTLs are defined as SNPs with abs(effectize) > 1e-8 in lasso results.  

2. Impute expression. 

We impute gene expression for our genotype data using expression models obtained from step 1. There are `r as.character(J)` genes with expression model. We imputed expression from genotypes using the expression predictors. 

3. Define and select regions

Next, the analysis is done at the “region” level. We used LDetect to define regions. 

4. Run ctwas_rss
We downsampled to 1/10 of SNPs to run ctwas_rss


```{r dirs}
library(ctwas)
library(data.table)
source('~/causalTWAS/causal-TWAS/code/qqplot.R')
outputdir <- "~/causalTWAS/simulations/simulation_ctwas_rss_20210117/"
runtag <- 'ukb-s80.45-adi'
```

# Parameter estimation results

Results: Each row shows parameter estimation results from 5 simulation runs with similar settings (i.e. pi1 and PVE for genes and SNPs). each row has two plots, one for gene pi1 estimation, one for enrichment (gene pi1/snp pi1). Results from each run were represented by one dot, dots with the same color come from the same run. horizontal dash lines: simulation truth,  `susietruth`, the truth in selected regions that were used to run susie iteractively (susieI). 


```{r param functions}
show_param <- function(phenofs, susieIfs, susieIfs2, thin = 1){
  pars <- do.call(rbind, lapply(phenofs, function(x) {load(x); 
    c(phenores$param$pve.gene.truth,
      phenores$param$pve.snp.truth,
      length(phenores$batch[[1]]$idx.cgene)/phenores$batch[[1]]$J,
      length(phenores$batch[[1]]$idx.cSNP)/phenores$batch[[1]]$M)}))
 
  colnames(pars) <- c("PVE.gene_truth", "PVE.SNP_truth", "pi1.gene_truth", "pi1.SNP_truth")
    
  param.s <- do.call(rbind, lapply(susieIfs, function(x) {load(x); c(tail(group_prior_rec[1,][!is.na(group_prior_rec[1,])], 1), tail(group_prior_rec[2,][!is.na(group_prior_rec[1,])], 1) * thin)}) )
  
  param.s.truth <- do.call(rbind, lapply(susieIfs2, function(x) {
    a <- fread(x, header = T);
    c(nrow(a[a$ifcausal == 1 & a$type == "gene" ])/ nrow(a[a$type == "gene"]),
      nrow(a[a$ifcausal == 1 & a$type == "SNP"])/ nrow(a[a$type == "SNP"]))
    }))
  
  pars.s <- cbind(param.s.truth, param.s)[, c(1,3,2,4)]
  colnames(pars.s) <-  paste(rep(c("pi1.gene_", "pi1.SNP_"), each = 2), c("susietruth", "susieI"), sep = "")
  
  df <- cbind(tags, format(pars, digits = 4), format(pars.s, digits =4))
  rownames(df) <- NULL
  return(df)
  
  # df %>% 
  # kable("html", escape = F) %>%
  # kable_styling("striped", full_width = F) %>%
  #  row_spec(c(1:5, 11:15), background = "#FEF3B9") %>%
  # scroll_box(width = "100%", height = "600px", fixed_thead = T)
}

plot_param <- function(df, ...){
  df <- apply(df[ , 2:ncol(df)], 2, function(x) as.numeric(x))
  st <-  cbind(df[,"pi1.gene_susietruth"], 1:nrow(df), 2 + 1:nrow(df)/nrow(df)/3)
  s <- cbind(df[,"pi1.gene_susieI"], 1:nrow(df), 3 + 1:nrow(df)/nrow(df)/3)
  t <- df[1,"pi1.gene_truth"]
  dfp <- rbind(st,s)
  plot(dfp[,3], dfp[,1], col = dfp[,2], pch = 19, ylab = "gene pi1", xaxt = "n", xlab="", xlim = c(0.8, 3.5), frame.plot=FALSE, ylim = c(0, max(dfp[,1],t) *1.05), ...)
  axis(side=1, at=1:2, labels = FALSE, tick = F)
  text(x=2:3, 0, labels = c( "susieI_truth", "susieI"), xpd = T, pos =1)
  abline(h=t, lty = 2, col= "salmon", lwd=1.5)
  grid()
  
  st <-  cbind(df[,"pi1.SNP_susietruth"], 1:nrow(df), 2 + 1:nrow(df)/nrow(df)/3)
  s <- cbind(df[,"pi1.SNP_susieI"], 1:nrow(df), 3 + 1:nrow(df)/nrow(df)/3)
  t <- df[1,"pi1.SNP_truth"]
  dfp <- rbind(st,s)
  plot(dfp[,3], dfp[,1], col = dfp[,2], pch = 19, ylab = "SNP pi1", xaxt = "n", xlab="", xlim = c(0.8, 3.5), frame.plot=FALSE, ylim = c(0, max(dfp[,1],t) *1.05), ...)
  axis(side=1, at=1:2, labels = FALSE, tick = F)
  text(x=2:3, 0, labels = c( "susieI_truth", "susieI"), xpd = T, pos =1)
  abline(h=t, lty = 2, col= "salmon", lwd=1.5)
  grid()
  
  df[,"pi1.SNP_susietruth"] <- 1
  st <-  cbind(df[,"pi1.gene_susietruth"]/df[,"pi1.SNP_susietruth"], 1:nrow(df), 2 + 1:nrow(df)/nrow(df)/3)
  s <- cbind(df[,"pi1.gene_susieI"]/df[,"pi1.SNP_susieI"], 1:nrow(df), 3 + 1:nrow(df)/nrow(df)/3)
  t <- df[1,"pi1.gene_truth"]/df[1,"pi1.SNP_truth"]
  dfp <- rbind(st,s)
  plot(dfp[,3], dfp[,1], col = dfp[,2], pch = 19, ylab = "Enrichment (gene/snp)", xaxt = "n", xlab="", xlim = c(0.8, 3.5),frame.plot=FALSE, ylim = c(0, min(max(dfp[,1],t) *1.05, 150)))
  axis(side=1, at=1:2, labels = FALSE, tick = F)
  text(x=2:3, 0, labels = c("susieI_truth", "susieI"), xpd = T, pos =1)
  abline(h= t, lty = 2, col= "darkgreen", lwd=1.5)
  grid()
}

```

## ctwas_rss (1)
* run parameters: downsample 1/10. 

```{r param1susie, fig.height=6, fig.width=8}
tags <- paste(rep(1:2, each = 5), 1:5, sep = "-")
phenofs <- paste0(outputdir, runtag, "_simu", tags, "-pheno.Rd")
susieIfs <- paste0(outputdir, runtag, "_simu", tags, "_config1.s2.susieIrssres.Rd")
susieIfs2 <- paste0(outputdir, runtag, "_simu",tags, "_config1.s2.susieIrss.txt")

df <- show_param(phenofs, susieIfs, susieIfs2, thin = 0.1)
par(mfrow = c(2,3))
plot_param(df[1:5,], main = "low power")
plot_param(df[6:10,], main = "high power")
```

