---
title: "Test SUSIE for regions near genes (500kb flanking)"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F)
```

```{r plot functions}
library(mr.ash.alpha)
library(data.table)
suppressMessages({library(plotly)})

plot_pip <- function(dt){
  pal <- c("salmon", "darkgreen")
  pal <- setNames(pal, c("Causal", "Non causal"))
  
  fig <- plot_ly(dt, x = ~ pos, y = ~ pip, type = 'scatter', mode = 'markers', color = ~ ifcausal, colors = pal, text = ~paste("Name: ", name), width = 600, height = 400)

  fig <- fig %>% layout( yaxis = list(title = "PIP"), plot_bgcolor='#ebf6fc')
  return(fig)
}

plot_all <- function(susieres, mrashres.gene, mrashres.snp, title){
  dt <- read.table(susieres , header =T)
  
  a <- read.table(mrashres.gene , header =T)
  a[a$ifcausal == 1, "ifcausal"] = "Causal"
  a[a$ifcausal == 0, "ifcausal"] = "Non causal"
  
  dt.gene <- merge(dt, a, by= "name")

  a <- read.table(mrashres.snp , header =T)
  a[a$ifcausal == 1, "ifcausal"] = "Causal"
  a[a$ifcausal == 0, "ifcausal"] = "Non causal"
  
  dt.snp <- merge(dt, a, by= "name")
  
  fig1 <- plot_pip(dt.gene)
  fig2 <- plot_pip(dt.snp)

  fig <- subplot(fig1, fig2, nrows= 2, shareX = T, titleY = T)
  
  fig <- fig %>% layout(title = title, autosize = F,  margin = 0.02,
    xaxis = list(title = "Genomic position"))
  
  fig <- fig %>% toWebGL()
  fig
}

```


```{r output dir}
simdatadir <- "~/causalTWAS/simulations/simulation_ashtest_20200616/"
outputdir <- "~/causalTWAS/simulations/simulation_susietest_20200616/"
mr.ash2dir <-  "~/causalTWAS/simulations/simulation_ashtest_20200616/"
```

## Simulation 5
```{r sim 5 setup}
tag <- "20200616-5"
tag2 <- "expr-res"
mrashres.gene <- paste0(mr.ash2dir, tag, "-mr.ash2s.", tag2, ".expr.txt")
mrashres.snp <- paste0(mr.ash2dir, tag, "-mr.ash2s.", tag2, ".snp.txt")
```

### SUSIE results

```{r sim 5, include= F, echo = F, eval = F, warning= F}
# displaying not right
resall <- Sys.glob(paste0(outputdir, tag, ".", tag2, ".L2*-susieres.txt"))
l <- htmltools::tagList()
res <- resall[1]
for (i in 1:8){
  res <- resall[i]
  name <- gsub(".*res-(.+)-susie.*", "\\1", res)
  l[[i]] <- plot_all(res, mrashres.gene, mrashres.snp, name)
}
l
```

```{r sim 5 examples,  out.width = "98%", out.height="95%", warning= F}
# displaying not right
res <- paste0(outputdir, tag, ".", tag2, ".L2-LL22NC03-86G7.1-susieres.txt")
name <- gsub(".*.L2-(.+)-susie.*", "\\1", res)
plot_all(res, mrashres.gene, mrashres.snp, name)
```
