---
title: "Summarize twas: chr17 to 22"
date: "`r Sys.Date()`"
output: workflowr::wflow_html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,  message= F, warning = F)
```
`

Run simulation 8 times for ukb chr 17 to chr 22 combined. SNPs are downsampled to 1/10, eQTLs defined by FUSION-TWAS (Adipose/GTEx) lasso effect size > 0 were kept, p= 86k, n = 20k.

```{r plot functions}
library(mr.ash.alpha)
library(data.table)
suppressMessages({library(plotly)})
library(tidyr)
library(plyr)
library(stringr)
library(kableExtra)
source("analysis/summarize_twas_plots.R")
```
```{r output dir}
simdatadir <- "~/causalTWAS/simulations/simulation_ashtest_20200721/"
outputdir <- "~/causalTWAS/simulations/simulation_ashtest_20200721/"
susiedir <- "~/causalTWAS/simulations/simulation_susietest_20200721/"
tags <- paste0('20200721-1-', c(2, 4:9))
tagglob <- '20200721-1-*'
tagextr <- '20200721-1-\\d+'
tag2s <- c('zeroes-es', 'zerose-es', 'lassoes-es','lassoes-se')
```

# Mr.ash2 parameter estimation

Results for 10 simulations runs, using different initialize and update strategy


## NULL; expr-snp; expr-snp 

```{r param 1}
show_param(tags, tag2s[1])
```

## NULL; snp-expr; expr-snp

```{r param 2}
show_param(tags, tag2s[2])
```

## lasso; expr-snp; expr-snp

```{r param 3}
show_param(tags, tag2s[3])
```

## lasso; expr-snp; snp-expr

```{r param 4}
show_param(tags, tag2s[4])
```

# Regional mr.ash2s PIP overview
Take simulation 1 (NULL; expr-snp; expr-snp) as examples. We use region size *500kb* and PIP cut off at *0.5* for SUSIE.

```{r rpip 1, fig.height=4, fig.width=10}
chrom = 18
f <- get_files(tag= "20200721-1-3" , tag2 = tag2s[1])
allchr <- read.table(f[["rpip"]], header = T)
a <- allchr[allchr["chrom"]==chrom,]
print(paste("plot for chr", chrom))
par(mar=c(5, 4, 4, 6) + 0.1)
with(a, plot(p0, rPIP, col ='salmon', xlab = "position", ylab= "Sum of PIP", type = 'h', lwd = 2))
par(new = T)
with(a, plot(p0, nCausal, pch =19, col = "darkgreen",axes = FALSE, bty = "n", xlab = "", ylab = ""))
axis(side = 4)
mtext(side = 4, line = 3, 'No. causal signals')
legend("topleft",
       legend=c("Mr.ASH PIP", "# Causal"),
       lty=c(1,0), pch=c(NA, 19), col=c("salmon", "darkgreen"))
grid()
```


# PIP calibration
We run 50 simulations and combine results.

## NULL; expr-snp; expr-snp 
```{r cali 1, fig.height=4, fig.width=14}
tag2 = "zeroes-es"
tags_ext <- Reduce(intersect, get_tags(tagglob, tagextr, tag2 = tag2)['gsusie'])
a <- caliPIP_plot(tags = tags_ext, tag2 = tag2)
```
```{r cali 1fdr, fig.height=4, fig.width=4}
a <- caliFDR_plot(tags = tags_ext, tag2 = tag2)
```

## Lasso; expr-snp; expr-snp 
```{r cali 2, fig.height=4, fig.width=14}
a <- caliPIP_plot(tags = tags_ext, tag2 = "lassoes-es")
```

```{r cali 2fdr, fig.height=4, fig.width=4}
a <- caliFDR_plot(tags = tags_ext, tag2 = "lassoes-es")
```

# PIP scatter plot
mr.ash2s PIP vs. susie PIP.

## NULL; expr-snp; expr-snp 
```{r sca 1, fig.height=3, fig.width=10}
scatter_plot_PIP(tags, tag2s[1])
```
## NULL; snp-expr; expr-snp

```{r sca 2, fig.height=3, fig.width=10}
scatter_plot_PIP(tags, tag2s[2])
```
## lasso; expr-snp; expr-snp

```{r sca 3, fig.height=3, fig.width=10}
scatter_plot_PIP(tags, tag2s[3])
```

## lasso; expr-snp; snp-expr

```{r sca 4, fig.height=3, fig.width=10}
scatter_plot_PIP(tags, tag2s[4])
```


# ROC curve

## NULL; expr-snp; expr-snp 
```{r roc 1, fig.height=4, fig.width=4}
tags <- paste0('20200721-1-', c(2,4:9))
ROC_plot(tags, tag2s[2])
```

## NULL; snp-expr; expr-snp

```{r roc 2, fig.height=4, fig.width=4}
ROC_plot(tags, tag2s[2])
```

## lasso; expr-snp; expr-snp

```{r roc 3, fig.height=4, fig.width=4}
ROC_plot(tags, tag2s[3])
```

## lasso; expr-snp; snp-expr

```{r roc 4, fig.height=4, fig.width=4}
ROC_plot(tags, tag2s[4])
```


# PIP vs p value

## Lasso; expr-snp; expr-snp 
```{r sca_p 1, fig.height=7, fig.width=7}
a <- scatter_plot_PIP_p(tags, "lassoes-es")
```

# Pretty plots
A) PIP calibration plot; B) FDP inflation of TWAS; C) Scatter plot of individual genes: PIPs vs. -log10-p values from TWAS.

We use SNP genotype data from chr 17 to chr 22 combined. These genomic regions represents 12.5% of the genome. SNPs are downsampled to 1/10 (randomly), eQTLs used in building the expression model were added back. We used lasso to train expression model using GTEx adipose tissue v7. In this simulation, we simulate phenotype with PVE explained by gene as 0.01, by SNP 0.05. pi1 for gene 0.05, pi1 for SNP 0.0025. 

For our method, we run mr.ash2s for both genes and SNPs (initiated with lasso). Then apply SUSIE for regions with sum of PIP > 0.5. This gives us PIP for individual genes.

For TWAS, we use linear regression for each gene and obtain p values.

```{r pretty data, include = F}
tag2 <- "lassoes-es"

# A)
tags_ext <- Reduce(intersect, get_tags(tagglob, tagextr, tag2 = tag2)['gsusie'])
res1 <- caliPIP_plot(tags = tags_ext, tag2 = tag2)

# B)
res2 <- caliFDR_plot(tags = tags_ext, tag2 = tag2)

# C)
tags <- paste0('20200721-1-', c(2,4:9))
res3 <- scatter_plot_PIP_p(tags, "lassoes-es")
```

```{r pretty plot, fig.height=3, fig.width=9}
par(mfrow=c(1,3))
cp_plot(res1$SUSIE.w0_PIP, res1$ifcausal, main = "SUSIE PIP calibration", col = "firebrick")
grid()
cp_plot(res2$FDR, res2$ifcausal, mode ="FDR", main = "TWAS FDP", col= "darkgreen")
grid()
plot(res3$TWAS, res3$SUSIE.w0, col = ifelse(res3$ifcausal=="Causal",  "salmon", "cyan3"), xlab = expression('TWAS -log'[10]*'(p)'), ylab = "SUSIE PIP",pch =19)
grid()
legend(9.5, 0.6, 
       legend = c("Causal", "Non causal"), 
      cex=0.9, pch=19, col=c("salmon", "cyan3"))
```


