---
title: "SNP-harmonization"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# SNP nomenclature 

I found this paper particularly useful for clarifying SNP nomenclature:

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6099125/


# File formats for SNP information

Different file formats have different names for describing alleles and coding alleles.

* `.bim` format: they use A1 and A2 alleles, they claim A1 is usually the minor allele, see here: https://www.cog-genomics.org/plink/1.9/formats#bim

* `.pvar` format: they use Alt and Ref alleles, see here: https://www.cog-genomics.org/plink/2.0/formats#pvar

* `.traw` format: see here: https://www.cog-genomics.org/plink/2.0/formats#traw
the Ref allele is counted (coded)

* Fusion summary stats files: They use A1, A2. A1 is the effect allele.

* Fusion weights files: They use .bim format for SNP information, so 5th column is Alt, 6th column is Ref. Effect size in `wgt.matrix` variable is defined using Alt as effect ellele (5th column in `snps` variable).

* SMR files: They use A1, A2. A1 is the effect allele. They also call A1 as Ref allele, A2 as Alt allele.

* ctwas format: `ctwas` package uses R package `pgenlibr` to read data. The Alt allele in .bim or .pvar files are counted, so the effect allele is Alt allele.

* plink association tests: see [here](http://zzz.bwh.harvard.edu/plink/anal.shtml). For the additive effects of SNPs, the direction of the regression coefficient represents the effect of each extra minor allele (i.e. a positive regression coefficient means that the minor allele increases risk/phenotype mean). If the --beta command is added along with --logistic, then the regression coefficients rather than the odds ratios will be returned.
NOTE:Elsewhere in plink, the term reference allele is sometimes used to refer to A1, i.e. the --reference-allele command can be used to specify which allele is A1. Note that in association testing, the odds ratios, etc are typically calculated with A2 as the actual reference allele (i.e. a positive OR means A1 increases risk relative to A2).

* GTEx summary statistics: The normalized effect size (NES) of the eQTLs is defined as the slope of the linear regression, and is computed as the effect of the alternative allele (ALT) relative to the reference allele (REF) in the human genome reference GRCh38/hg38 (i.e., the eQTL effect allele is the ALT allele). See [here](https://www.gtexportal.org/home/documentationPage).

* GEMMA: "chr    rs  ps  n_miss  allel1  allel0  af  beta    se  l_remle p_wald". The 11 columns are: chromosome, SNP ID, basepair position, number of missing values for a given SNP, the effect (coded) allele, the other allele, frequency of the effect allele, effect size, standard error, lambda and p-value ([http://www.xzlab.org/software.html]).

# Consistency between GTEx v7 and UKB data

* We check if SNP information provided by FUSION GTEx v7 trained and UKB SNP information are consistent.

UKBiobank SNP information were generated by plink2, we only selected variants with MAF > 0.05 variants and missing genotype rate < 0.05.

eQTL information were obtained from FUSION pre-trained models. We used all SNPs with weights != 0 in lasso trained models.

```{r consistent}
library(ctwas)
library(plyr)
ld_pgenfn = "~/causalTWAS/ukbiobank/ukb_pgen_s40.22/ukb-s40.22.2_pgenfs.txt"
eqtlfn = "~/causalTWAS/fusion_weights/Adipose_Subcutaneous.lasso.eqtl.txt"

eqtl = read.table(eqtlfn, header = T)
colnames(eqtl)[1:6] <- c("chrom", "id", "cm", "pos", "alt", "ref")
ld_pgenfs <- read.table(ld_pgenfn, header = F, stringsAsFactors = F)[,1]
ld_pvarfs <- sapply(ld_pgenfs, prep_pvar, outputdir = "~/temp/")

i <- 1 # chrom 1 as an example
pvar <- read_pvar(ld_pvarfs[i])

cat("number of eQTLs on chromosome ", i,  "(by lasso): ", nrow(eqtl[eqtl$chrom == i,]))

a <- join(pvar, eqtl, type = "inner", by = c("id", "pos"))
cat("number of eQTLs on chromosome ", i,  "(by lasso, shared with UKB): ", nrow(a))

b <- join(pvar, eqtl, type = "inner")
cat("number of eQTLs on chromosome ", i,  "(by lasso, shared with UKB) that match: ", nrow(b))

cat("number of eQTLs on chromosome ", i,  "(by lasso, shared with UKB) that match after reverse alt and ref: ",
    which(a[!(a$id %in% b$id), 5] != a[!(a$id %in% b$id), 8]))

cat("number of eQTLs on chromosome ", i,  "(by lasso, shared with UKB) that match after flip: ",
sum(a[!(a$id %in% b$id), 5] != a[!(a$id %in% b$id), 8]))

cat("number of eQTLs on chromosome ", i,  "(by lasso, shared with UKB) that are strand ambiguous: ",
nrow(a[(a$alt=="A" & a$ref=="T") | (a$alt=="T" & a$ref=="A") | (a$alt=="C" & a$ref=="G") | (a$alt=="G" & a$ref=="C"),]))
```

# Procedures to harmonize SNPs

See here about some instructions for merging two datasets: 
https://zzz.bwh.harvard.edu/plink/dataman.shtml#flip

It is assumed that reported two alleles are on the same strand, but it can be either positive strand or negative strand.

* FUSION filters out all strand ambiguous SNPs (i.e. AT, TA, CG, GC alleles) and flipped strand unambiguous SNPs if there is a match after flipping. It also switch alt ref SNPs if there is a match after switching. 

* SMR switch alt ref SNPs if there is a match after switching. No flipping.

* ctwas currently (v.0.1.11) does not harmonize data and assumes the alleles were harmonized before running ctwas, so they do not need to be flipped or switched.


