---
title: "SNP-harmonization"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# SNP nomenclature 

I found this paper particularly useful for clarifying SNP nomenclature:

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6099125/


# File formats for SNP information

Different file formats have different names for describing alleles and coding alleles.

* `.bim` format: they use A1 and A2 alleles, they claim A1 is usually the minor allele, see here: https://www.cog-genomics.org/plink/1.9/formats#bim

* `.pvar` format: they use Alt and Ref alleles, see here: https://www.cog-genomics.org/plink/2.0/formats#pvar

* `.traw` format: see here: https://www.cog-genomics.org/plink/2.0/formats#traw
the Ref allele is counted (coded)

* Fusion summary stats files: They use A1, A2. A1 is the effect allele.

* Fusion weights files: They use .bim format for SNP information, so 5th column is Alt, 6th column is Ref.

* SMR files: They use A1, A2. A1 is the effect allele. They also call A1 as Ref allele, A2 as Alt allele.

* ctwas format: `ctwas` package uses R package `pgenlibr` to read data. The Alt allele in .bim or .pvar files are counted, so the effect allele is Alt allele.

# Consistency between GTEx v7 and UKB data

* We check if SNP information provided by FUSION GTEx v7 trained and UKB SNP information are consistent.

UKBiobank SNP information were generated by plink2, we only selected variants with MAF > 0.05 variants and missing genotype rate < 0.05.

eQTL information were obtained from FUSION pre-trained models. We used all SNPs with weights != 0 in lasso trained models.

```{r consistent}
library(ctwas)
library(plyr)
ld_pgenfn = "~/causalTWAS/ukbiobank/ukb_pgen_s40.22/ukb-s40.22.2_pgenfs.txt"
eqtlfn = "~/causalTWAS/fusion_weights/Adipose_Subcutaneous.lasso.eqtl.txt"

eqtl = read.table(eqtlfn, header = T)
colnames(eqtl)[1:6] <- c("chrom", "id", "cm", "pos", "alt", "ref")
ld_pgenfs <- read.table(ld_pgenfn, header = F, stringsAsFactors = F)[,1]
ld_pvarfs <- sapply(ld_pgenfs, prep_pvar, outputdir = "~/temp/")

i <- 1 # chrom 1 as an example
pvar <- read_pvar(ld_pvarfs[i])

cat("number of eQTLs on chromosome ", i,  "(by lasso): ", nrow(eqtl[eqtl$chrom == i,]))

a <- join(pvar, eqtl, type = "inner", by = c("id", "pos"))
cat("number of eQTLs on chromosome ", i,  "(by lasso, shared with UKB): ", nrow(a))

b <- join(pvar, eqtl, type = "inner")
cat("number of eQTLs on chromosome ", i,  "(by lasso, shared with UKB) that match: ", nrow(b))

cat("number of eQTLs on chromosome ", i,  "(by lasso, shared with UKB) that match after reverse alt and ref: ",
    which(a[!(a$id %in% b$id), 5] != a[!(a$id %in% b$id), 8]))

cat("number of eQTLs on chromosome ", i,  "(by lasso, shared with UKB) that match after flip: ",
sum(a[!(a$id %in% b$id), 5] != a[!(a$id %in% b$id), 8]))

cat("number of eQTLs on chromosome ", i,  "(by lasso, shared with UKB) that are strand ambiguous: ",
nrow(a[(a$alt=="A" & a$ref=="T") | (a$alt=="T" & a$ref=="A") | (a$alt=="C" & a$ref=="G") | (a$alt=="G" & a$ref=="C"),]))
```

# Procedures to harmonize SNPs

See here about some instructions for merging two datasets: 
https://zzz.bwh.harvard.edu/plink/dataman.shtml#flip

It is assumed that reported two alleles are on the same strand, but it can be either positive strand or negative strand.

* FUSION filters out all strand ambiguous SNPs (i.e. AT, TA, CG, GC alleles) and flipped strand unambiguous SNPs if there is a match after flipping. It also switch alt ref SNPs if there is a match after switching. 

* SMR switch alt ref SNPs if there is a match after switching. No flipping.

* ctwas currently (v.0.1.11) does not harmonize data and assumes the alleles were harmonized before running ctwas, so they do not need to be flipped or switched.


