---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Analysis description

```{r parameters}
n.ori <- 40000 # number of samples
n <- 100
p <-100 # number of SNPs
J <- 8021 # number of genes
```

## Data

  * **GWAS summary statistics** we simulated summary statistics data with different causal gene/SNP proportion and PVE. To simulate this data, we need the following:
      * **genotype data** we used is from UKB biobank, randomly selecting `r as.character(n.ori)` samples. We then filtered samples.
      The columns selected for filtering samples are as follows:
        * sex (31)
        * UK Biobank assessment centre (54)
        * age (21022)
        * genetic ethnic grouping (22006)
        * genetic sex (22001)
        * genotype measurement batch (22000)
        * missingness (22005)
        * genetic PCs (22009-0.1 - 22009-0.40)
        * genetic relatedness pairing (22011)
        * genetic kinship (22021)
        * outliers (22027)
        
        We filtered the samples based on the following criteria:  
        1. Remove all rows in which one or more of the values are missing, aside from the in the "outlier" and "relatedness_genetic" columns. This step removes any samples that are not marked as being "White British".  
        2. Remove rows with mismatches between self-reported and genetic sex  
        3. Remove "missingness" and "heterozygosity" outliers as defined by UK Biobank. 
        4. Remove any individuals have at least one relative  
        5. Remove any individuals that have close relatives identified from the "relatendess" calculations that weren't already identified using the kinship calculations.
        
        This ended up with **n = `r as.character(n)` samples**.  We use SNP genotype data from chr 1 to chr 22 combined from UKB. We select SNPs with minor allele frequency > 0.05 and SNPs with at least 95% genotyping rate (5% missing). There are **total = `r as.character(p)` SNPs** passed these filters and go into our analysis.  
    * **Expression models** The one we used in this analysis is GTEx Adipose tissue v7 dataset. This dataset contains ~ 380 samples, **`r as.character(J)` genes** with expression model. [FUSION/TWAS](http://gusevlab.org/projects/fusion/) were used to train expression model and we used their lasso results. SNPs included in eQTL anlaysis are restricted to cis-locus 500kb on either side of the gene boundary. eQTLs are defined as SNPs with abs(effectize) > 1e-8 in lasso results. 
      
      To simulate phenotype data, first we impute gene expression based on expression models, then we set gene/SNP pi1 and PVE, get rough effect size for causal SNPs and genes and simulate phenotype under the sparse model with spike and slab prior.
      Then we performed GWAS for all SNPs and get z scores for each by univariate linear regression. 
  
  * **LD genotype reference**
  We used the genotype of 2k samples from UKbiobank (randomly selected from the samples used in simulations) to serve as the LD reference. 
  
  * **Expression models**  
  We used GTEx Adipose tissue v7 dataset, the same as used for simulating phenotypes.

## Analysis

### ctwas

1. Get z scores for gene expression. 
We used expression models and LD reference to get z scores for gene expression.

2. Run ctwas_rss
We used LDetect to define regions. `ctwas_rss` algorithm first runs on all regions to get rough estimate for gene and SNP prior. Then run on small regions (having small probablities of having > 1 causal signals based on rough estimates) to get more accurate estimate. To lower computational burden, we downsampled SNPs (0.1), estimate parameters and convert back to orginal scale. Lastly, run susie with given L for all regions and for all genes and SNPs using estimated prior and prior variance.

